<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>SFEWorkshop2024 • SFEWorkshop2024</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="SFEWorkshop2024">
<meta name="description" content="This workshop showcases new features in SpatialFeatureExperiment and Voyager, including support for transcript spot geometries in rowGeometries, plotting transcript spots, aggregating transcript spots with a spatial grid or any geometry, aggregating cell level data with a spatial grid or any geometry, splitting SFE objects with geometries, and improved support for images and affine transformations.">
<meta property="og:description" content="This workshop showcases new features in SpatialFeatureExperiment and Voyager, including support for transcript spot geometries in rowGeometries, plotting transcript spots, aggregating transcript spots with a spatial grid or any geometry, aggregating cell level data with a spatial grid or any geometry, splitting SFE objects with geometries, and improved support for images and affine transformations.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">SFEWorkshop2024</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/workshop.html">Workshop</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/lambdamoses/SFEWorkshop2024/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="bioc2024-sfe-and-voyager-workshop">Bioc2024 SFE and Voyager workshop<a class="anchor" aria-label="anchor" href="#bioc2024-sfe-and-voyager-workshop"></a>
</h1></div>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<div class="section level3">
<h3 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h3>
<p>Exploratory spatial data analysis (ESDA) can be a powerful approach to understanding single-cell genomics datasets, but it is not yet part of standard data analysis workflows. In particular, geospatial analyses, which have been developed and refined for decades, have yet to be fully adapted and applied to spatial single-cell analysis. We introduce the <code>Voyager</code> platform, which systematically brings the geospatial ESDA tradition to (spatial) -omics, with local, bivariate, and multivariate spatial methods not yet commonly applied to spatial -omics, united by a uniform user interface. Using <code>Voyager</code>, we showcase biological insights that can be derived with its methods, such as biologically relevant negative spatial autocorrelation. Underlying <code>Voyager</code> is the <code>SpatialFeatureExperiment</code> (SFE) data structure, which combines Simple Feature with <code>SingleCellExperiment</code> and <code>AnnData</code> to represent and operate on geometries bundled with gene expression data. Voyager has comprehensive tutorials demonstrating ESDA built on GitHub Actions to ensure reproducibility and scalability, using data from popular commercial technologies. Voyager is implemented in both R/Bioconductor and Python/PyPI, and features compatibility tests to ensure that both implementations return consistent results.</p>
<p>This workshop focuses on new features of SFE and Voyager implemented in the past year. Earlier features are described in more details <a href="https://www.biorxiv.org/content/10.1101/2023.07.20.549945v1?ct=" class="external-link">this preprint</a> and in the <a href="https://lambdamoses.github.io/VoyagerWorkshop/" class="external-link">2023 Voyager workshop</a>. The preprint will soon be updated to include the new functionalities prior to submission. The new features fall under these areas:</p>
<ol style="list-style-type: decimal">
<li>Expanded support for images. A new image class <code>BioFormatsImage</code> has been implemented to support OME-TIFF from Xenium and <code>ExtImage</code> class to support image processing with the <code>EBImage</code> package.</li>
<li>Support for transcript spot geometries. Coordinates from smFISH-based technologies are reformatted for faster reading next time, and is stored in the <code>rowGeometries</code> field of the SFE object. As simple features, geometric operations from <code>sf</code> can be applied.</li>
<li>Support for images and transcript spots led to improved functions to read output from commercial technologies: Vizgen MERFISH, Xenium, CosMX. We’re working on VisiumHD.</li>
<li>Expanded support for affine transformations. Affine transformation can be performed with a transformation matrix, and all images and geometries are transformed while remaining aligned.</li>
<li>New spatial operations on SFE objects: spatial aggregation (cell and transcript spot levels) and splitting by geometries.</li>
</ol>
</div>
<div class="section level3">
<h3 id="pre-requisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#pre-requisites"></a>
</h3>
<p>To understand the workshop material, you are expected to have:</p>
<ul>
<li>Intermediate knowledge of R syntax</li>
<li>At least basic familiarity with Tidyverse and <code>ggplot2</code>
</li>
<li>Familiarity with <code>sf</code> is encouraged. Read <a href="https://r.geocompx.org/" class="external-link">Geocomputing with R</a> for an intro to <code>sf</code> and geospatial data.</li>
<li>It would be helpful to take a look at <a href="https://lambdamoses.github.io/VoyagerWorkshop/" class="external-link">last year’s Voyager workshop</a>; this year I’m not covering the math behind ESDA, in order to focus on the new functionalities.</li>
</ul>
</div>
<div class="section level3">
<h3 id="participation">Participation<a class="anchor" aria-label="anchor" href="#participation"></a>
</h3>
<p>This workshop consists of hands on demos, exercises, and Q&amp;A.</p>
</div>
<div class="section level3">
<h3 id="r--bioconductor-packages-used">
<em>R</em> / <em>Bioconductor</em> packages used<a class="anchor" aria-label="anchor" href="#r--bioconductor-packages-used"></a>
</h3>
<ul>
<li><a href="https://pachterlab.github.io/voyager/" class="external-link"><code>Voyager</code></a></li>
<li><a href="https://pachterlab.github.io/SpatialFeatureExperiment/" class="external-link"><code>SpatialFeatureExperiment</code></a></li>
<li><a href="https://bioconductor.org/packages/release/data/experiment/html/SFEData.html" class="external-link"><code>SFEData</code></a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/scater.html" class="external-link"><code>scater</code></a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/scran.html" class="external-link"><code>scran</code></a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html" class="external-link"><code>SingleCellExperiment</code></a></li>
</ul>
</div>
<div class="section level3">
<h3 id="time-outline">Time outline<a class="anchor" aria-label="anchor" href="#time-outline"></a>
</h3>
<p>Total: 90 minutes</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Introduction and setup</td>
<td>15m</td>
</tr>
<tr class="even">
<td>Read data into R as SFE</td>
<td>15m</td>
</tr>
<tr class="odd">
<td>Getting, setting, and plotting SFE fields</td>
<td>15m</td>
</tr>
<tr class="even">
<td>Images in SFE</td>
<td>15m</td>
</tr>
<tr class="odd">
<td>Spatial operations of SFE</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Aggregate and ESDA at different scales</td>
<td>10m</td>
</tr>
<tr class="odd">
<td>Split and ESDA across samples</td>
<td>10m</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="workshop-goals-and-objectives">Workshop goals and objectives<a class="anchor" aria-label="anchor" href="#workshop-goals-and-objectives"></a>
</h3>
<ul>
<li>Learn basics of SFE and Voyager ESDA</li>
<li>Learn to work with images and transcript spots in SFE</li>
<li>Discussion on ESDA ideas for different lengths scales and samples</li>
</ul>
<div class="section level4">
<h4 id="what-you-will-learn">What you will learn<a class="anchor" aria-label="anchor" href="#what-you-will-learn"></a>
</h4>
<ul>
<li>Use and operate on SFE objects</li>
<li>Spatial data visualization</li>
<li>Geometric operations</li>
<li>Create spatial neighborhood graphs</li>
<li>Run spatial analyses on different fields of SFE objects</li>
<li>Visualize spatial analysis results</li>
</ul>
</div>
<div class="section level4">
<h4 id="what-you-will-not-learn">What you will <em>not</em> learn<a class="anchor" aria-label="anchor" href="#what-you-will-not-learn"></a>
</h4>
<ul>
<li>Technologies to collect spatial transcriptomics data</li>
<li>Data integration across multiple samples</li>
<li>Spatial multi-omics</li>
</ul>
</div>
</div>
</div>
<div class="section level2">
<h2 id="to-use-the-docker-image">To use the Docker image<a class="anchor" aria-label="anchor" href="#to-use-the-docker-image"></a>
</h2>
<p>This workshop can be run remotely on the <a href="https://workshop.bioconductor.org/" class="external-link">Bioconductor Workshop Galaxy</a>. Alternatively, it can be run locally with the Docker image:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">-e</span> PASSWORD=<span class="op">&lt;</span>choose_a_password_for_rstudio<span class="op">&gt;</span> -p 8787:8787 ghcr.io/lambdamoses/sfeworkshop2024</span></code></pre></div>
<p>Once running, navigate to <a href="http://localhost:8787/" class="external-link uri">http://localhost:8787/</a> and then log in with <code>rstudio</code>:<code>yourchosenpassword</code>.</p>
<p>The required packages of the appropriate version as of July 2024 have been pre-installed on the Workshop Galaxy and in the Docker image. Devel versions of SFE, Voyager, and Bioconductor as of July 2024 are used here.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/lambdamoses/SFEWorkshop2024/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/lambdamoses/SFEWorkshop2024/issues/new/choose" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing SFEWorkshop2024</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Lambda Moses <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-7092-9427" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Alik Huseynov <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-1438-4389" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Lior Pachter <br><small class="roles"> Author, thesis advisor </small> <a href="https://orcid.org/0000-0002-9164-6231" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lambda Moses, Alik Huseynov, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
