<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Images and transcript spots in SFE • SFEWorkshop2024</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Images and transcript spots in SFE">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SFEWorkshop2024</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/workshop.html">Workshop</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/lambdamoses/SFEWorkshop2024/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Images and transcript spots in SFE</h1>
                        <h4 data-toc-skip class="author">Lambda
Moses</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:dl3764@columbia.edu" class="email">dl3764@columbia.edu</a>
      
                              <h4 data-toc-skip class="author">Alik
Huseynov</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:alikhuseyno@gmail.com" class="email">alikhuseyno@gmail.com</a>
      
                              <h4 data-toc-skip class="author">Lior
Pachter</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:lpachter@caltech.edu" class="email">lpachter@caltech.edu</a>
      
                  
            <h4 data-toc-skip class="date">2024-07-23</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/lambdamoses/SFEWorkshop2024/blob/devel/vignettes/workshop.Rmd" class="external-link"><code>vignettes/workshop.Rmd</code></a></small>
      <div class="d-none name"><code>workshop.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="spatialfeatureexperiment">
<code>SpatialFeatureExperiment</code><a class="anchor" aria-label="anchor" href="#spatialfeatureexperiment"></a>
</h2>
<p><code>SpatialFeatureExperiment</code> (SFE) is a new <a href="https://adv-r.hadley.nz/s4.html" class="external-link">S4</a> class built on top of <a href="https://bioconductor.org/packages/release/bioc/html/SpatialExperiment.html" class="external-link"><code>SpatialExperiment</code></a>
(SPE). SFE incorporates geometries and geometric operations with the <a href="https://cran.r-project.org/web/packages/sf/index.html" class="external-link"><code>sf</code></a>
package. Examples of supported geometries are Visium spots represented
with polygons corresponding to their size, cell or nuclei segmentation
polygons, tissue boundary polygons, pathologist annotation of
histological regions, and transcript spots of genes. Using
<code>sf</code>, <code>SpatialFeatureExperiment</code> leverages the
GEOS C++ library underlying <code>sf</code> for geometry operations,
including algorithms for for determining whether geometries intersect,
finding intersection geometries, buffering geometries with margins, etc.
A schematic of the SFE object is shown below:</p>
<p><img src="sfe_schematics.png" alt="SpatialFeatureExperiment expands on SpatialExperiment by adding column, row, and annotation geometries and spatial graphs. This is explained in detail in the following paragraphs." width="100%"></p>
<p>Below is a list of SFE features that extend the SPE object:</p>
<ul>
<li>
<code>colGeometries</code> are <code>sf</code> data frames
associated with the entities that correspond to columns of the gene
count matrix, such as Visium spots or cells. The geometries in the
<code>sf</code> data frames can be Visium spot centroids, Visium spot
polygons, or for datasets with single cell resolution, cell or nuclei
segmentations. Multiple <code>colGeometries</code> can be stored in the
same SFE object, such as one for cell segmentation and another for
nuclei segmentation. There can be non-spatial, attribute columns in a
<code>colGeometry</code> rather than <code>colData</code>, because the
<code>sf</code> class allows users to specify how attributes relate to
geometries, such as “constant”, “aggregate”, and “identity”. See the
<code>agr</code> argument of the <a href="https://r-spatial.github.io/sf/reference/sf.html" class="external-link"><code>st_sf</code>
documentation</a>.</li>
<li>
<code>colGraphs</code> are spatial neighborhood graphs of cells or
spots. The graphs have class <code>listw</code> (<code>spdep</code>
package), and the <code>colPairs</code> of
<code>SingleCellExperiment</code> was not used so no conversion is
necessary to use the numerous spatial dependency functions from
<code>spdep</code>, such as those for Moran’s I, Geary’s C, Getis-Ord
Gi*, LOSH, etc. Conversion is also not needed for other classical
spatial statistics packages such as <code>spatialreg</code> and
<code>adespatial</code>.</li>
<li>
<code>rowGeometries</code> are similar to
<code>colGeometries</code>, but support entities that correspond to rows
of the gene count matrix, such as genes. As we shall see below, a use
case is to store transcript spots for each gene in smFISH or in situ
sequencing based datasets.</li>
<li>
<code>rowGraphs</code> are similar to <code>colGraphs</code>. A
potential use case may be spatial colocalization of transcripts of
different genes.</li>
<li>
<code>annotGeometries</code> are <code>sf</code> data frames
associated with the dataset but not directly with the gene count matrix,
such as tissue boundaries, histological regions, cell or nuclei
segmentation in Visium datasets, and etc. These geometries are stored in
this object to facilitate plotting and using <code>sf</code> for
operations such as to find the number of nuclei in each Visium spot and
which histological regions each Visium spot intersects. Unlike
<code>colGeometries</code> and <code>rowGeometries</code>, the number of
rows in the <code>sf</code> data frames in <code>annotGeometries</code>
is not constrained by the dimension of the gene count matrix and can be
arbitrary.</li>
<li>
<code>annotGraphs</code> are similar to <code>colGraphs</code> and
<code>rowGraphs</code>, but are for entities not directly associated
with the gene count matrix, such as spatial neighborhood graphs for
nuclei in Visium datasets, or other objects like myofibers. These graphs
are relevant to <code>spdep</code> analyses of attributes of these
geometries such as spatial autocorrelation in morphological metrics of
myofibers and nuclei. With geometry operations with <code>sf</code>,
these attributes and results of analyses of these attributes
(e.g. spatial regions defined by the attributes) may be related back to
gene expression.</li>
<li>
<code>localResults</code> are similar to <a href="https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html#3_Adding_low-dimensional_representations" class="external-link"><code>reducedDims</code>
in <code>SingleCellExperiment</code></a>, but stores results from
univariate and bivariate local spatial analysis results, such as from <a href="https://r-spatial.github.io/spdep/reference/localmoran.html" class="external-link"><code>localmoran</code></a>,
<a href="https://r-spatial.github.io/spdep/reference/localG.html" class="external-link">Getis-Ord
Gi*</a>, and <a href="https://r-spatial.github.io/spdep/reference/LOSH.html" class="external-link">local
spatial heteroscedasticity (LOSH)</a>. Unlike in
<code>reducedDims</code>, for each type of results (type is the type of
analysis such as Getis-Ord Gi*), each feature (e.g. gene) or pair of
features for which the analysis is performed has its own results. The
local spatial analyses can also be performed for attributes of
<code>colGeometries</code> and <code>annotGeometries</code> in addition
to gene expression and <code>colData</code>. Results of multivariate
spatial analysis such as <a href="https://cran.r-project.org/web/packages/adespatial/vignettes/tutorial.html#multispati-analysis" class="external-link">MULTISPATI
PCA</a> can be stored in <code>reducedDims</code>.</li>
<li>
<code>imgData</code> store images associated with the dataset. This
field is inherited from SPE, but SFE has extended the image
functionalities so images are not loaded into memory unless
necessary.</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-spatial.github.io/sf/" class="external-link">sf</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://fs.r-lib.org" class="external-link">fs</a></span><span class="op">)</span> <span class="co"># I put it after sf intentionally</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rspatial.org/" class="external-link">terra</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/voyager" class="external-link">Voyager</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/aoles/EBImage" class="external-link">EBImage</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://scales.r-lib.org" class="external-link">scales</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/apache/arrow/" class="external-link">arrow</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/const-ae/sparseMatrixStats" class="external-link">sparseMatrixStats</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://adriancorrendo.github.io/metrica/" class="external-link">metrica</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The material here is either taken from the documentation websites of
<a href="https://pachterlab.github.io/SpatialFeatureExperiment/" class="external-link"><code>SpatialFeatureExperiment</code></a>
and <a href="https://pachterlab.github.io/voyager/" class="external-link"><code>Voyager</code></a>
(release version; devel version: <a href="https://pachterlab.github.io/SpatialFeatureExperiment/dev/" class="external-link">SFE</a>,
<a href="https://pachterlab.github.io/voyager/dev/" class="external-link"><code>Voyager</code></a>)
or will be added to them. Some topics are covered in more details on the
documentation websites. A rendered version of the workshop can be seen
<a href="https://lambdamoses.github.io/SFEWorkshop2024/">here</a>.</p>
</div>
<div class="section level2">
<h2 id="sfe-object-construction">SFE object construction<a class="anchor" aria-label="anchor" href="#sfe-object-construction"></a>
</h2>
<div class="section level3">
<h3 id="from-scratch">From scratch<a class="anchor" aria-label="anchor" href="#from-scratch"></a>
</h3>
<p>An SFE object can be constructed from scratch with the assay matrices
and metadata. In this toy example, <code>dgCMatrix</code> is used, but
since SFE inherits from <code>SingleCellExperiment</code> (SCE), other
types of arrays supported by SCE such as delayed arrays should also
work.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visium barcode location from Space Ranger</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"visium_row_col"</span><span class="op">)</span></span>
<span><span class="va">coords1</span> <span class="op">&lt;-</span> <span class="va">visium_row_col</span><span class="op">[</span><span class="va">visium_row_col</span><span class="op">$</span><span class="va">col</span> <span class="op">&lt;</span> <span class="fl">6</span> <span class="op">&amp;</span> <span class="va">visium_row_col</span><span class="op">$</span><span class="va">row</span> <span class="op">&lt;</span> <span class="fl">6</span>,<span class="op">]</span></span>
<span><span class="va">coords1</span><span class="op">$</span><span class="va">row</span> <span class="op">&lt;-</span> <span class="va">coords1</span><span class="op">$</span><span class="va">row</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Random toy sparse matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span><span class="va">col_inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">13</span>, <span class="fl">13</span><span class="op">)</span></span>
<span><span class="va">row_inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">13</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">values</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">13</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/sparseMatrix.html" class="external-link">sparseMatrix</a></span><span class="op">(</span>i <span class="op">=</span> <span class="va">row_inds</span>, j <span class="op">=</span> <span class="va">col_inds</span>, x <span class="op">=</span> <span class="va">values</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">coords1</span><span class="op">$</span><span class="va">barcode</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">LETTERS</span>, <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p>That should be sufficient to create an SPE object, and an SFE object,
even though no <code>sf</code> data frame was constructed for the
geometries. The constructor behaves similarly to the SPE constructor.
The centroid coordinates of the Visium spots in the toy example can be
converted into spot polygons with the <code>spotDiameter</code>
argument. Spot diameter in pixels in full resolution image can be found
in the <code>scalefactors_json.json</code> file in Space Ranger
output.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/SpatialFeatureExperiment.html" class="external-link">SpatialFeatureExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">mat</span><span class="op">)</span>, colData <span class="op">=</span> <span class="va">coords1</span>,</span>
<span>                                spatialCoordsNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"col"</span>, <span class="st">"row"</span><span class="op">)</span>,</span>
<span>                                spotDiameter <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="space-ranger-output">Space Ranger output<a class="anchor" aria-label="anchor" href="#space-ranger-output"></a>
</h3>
<p>Space Ranger output can be read in a similar manner as in
<code>SpatialExperiment</code>; the returned SFE object has the
<code>spotPoly</code> column geometry for the spot polygons. If the
filtered matrix is read in, then a column graph called
<code>visium</code> will also be present, for the spatial neighborhood
graph of the Visium spots on tissue. The graph is not computed if all
spots are read in regardless of whether they are on tissue.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"SpatialFeatureExperiment"</span><span class="op">)</span></span>
<span><span class="va">sample_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample01"</span>, <span class="st">"sample02"</span><span class="op">)</span></span>
<span><span class="op">(</span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir</span>, <span class="va">sample_ids</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/usr/local/lib/R/site-library/SpatialFeatureExperiment/extdata/sample01"</span></span>
<span><span class="co">#&gt; [2] "/usr/local/lib/R/site-library/SpatialFeatureExperiment/extdata/sample02"</span></span></code></pre></div>
<p>Inside the <code>outs</code> directory:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"outs"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "filtered_feature_bc_matrix" "spatial"</span></span></code></pre></div>
<p>There should also be <code>raw_feature_bc_matrix</code> though this
toy example only has the filtered matrix.</p>
<p>Inside the matrix directory:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"outs"</span>, <span class="st">"filtered_feature_bc_matrix"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "barcodes.tsv" "features.tsv" "matrix.mtx"</span></span></code></pre></div>
<p>Inside the <code>spatial</code> directory:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"outs"</span>, <span class="st">"spatial"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "aligned_fiducials.jpg"              "barcode_fluorescence_intensity.csv"</span></span>
<span><span class="co">#&gt; [3] "detected_tissue_image.jpg"          "scalefactors_json.json"            </span></span>
<span><span class="co">#&gt; [5] "spatial_enrichment.csv"             "tissue_hires_image.png"            </span></span>
<span><span class="co">#&gt; [7] "tissue_lowres_image.png"            "tissue_positions.csv"</span></span></code></pre></div>
<p>Not all Visium datasets have all the files here. The
<code>barcode_fluorescence_intensity.csv</code> file is only present for
datasets with fluorescent imaging rather than bright field H&amp;E.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/read10xVisiumSFE.html" class="external-link">read10xVisiumSFE</a></span><span class="op">(</span><span class="va">samples</span>, sample_id <span class="op">=</span> <span class="va">sample_ids</span>, type <span class="op">=</span> <span class="st">"sparse"</span>, </span>
<span>                          data <span class="op">=</span> <span class="st">"filtered"</span>, images <span class="op">=</span> <span class="st">"hires"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; 10X Visium data will be loaded: sample01</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Adding spatial neighborhood graph to sample01</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; 10X Visium data will be loaded: sample02</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Adding spatial neighborhood graph to sample02</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 5 25 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(5): ENSG00000014257 ENSG00000142515 ENSG00000263639</span></span>
<span><span class="co">#&gt;   ENSG00000163810 ENSG00000149591</span></span>
<span><span class="co">#&gt; rowData names(14): symbol Feature.Type ...</span></span>
<span><span class="co">#&gt;   Median.Normalized.Average.Counts_sample02</span></span>
<span><span class="co">#&gt;   Barcodes.Detected.per.Feature_sample02</span></span>
<span><span class="co">#&gt; colnames(25): GTGGCGTGCACCAGAG-1 GGTCCCATAACATAGA-1 ...</span></span>
<span><span class="co">#&gt;   TGCAATTTGGGCACGG-1 ATGCCAATCGCTCTGC-1</span></span>
<span><span class="co">#&gt; colData names(10): in_tissue array_row ... channel3_mean channel3_stdev</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : pxl_col_in_fullres pxl_row_in_fullres</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixel</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01: col: visium</span></span>
<span><span class="co">#&gt; sample02: col: visium</span></span></code></pre></div>
<p>This workshop focuses on imaging based data from Xenium and MERFISH;
see <a href="https://pachterlab.github.io/SpatialFeatureExperiment" class="external-link">the
SFE vignette</a> for more details about Visium. This example here is for
old fashioned Visium; we are working on a read function and vignette for
higher resolution <a href="https://www.10xgenomics.com/products/visium-hd-spatial-gene-expression" class="external-link">VisiumHD</a>
data.</p>
</div>
<div class="section level3">
<h3 id="vizgen-merfish-output">Vizgen MERFISH output<a class="anchor" aria-label="anchor" href="#vizgen-merfish-output"></a>
</h3>
<p>The commercialized MERFISH from Vizgen has a standard output format,
that can be read into SFE with <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readVizgen.html" class="external-link">readVizgen()</a></code>. Because the
cell segmentation from each field of view (FOV) has a separate HDF5 file
and a MERFISH dataset can have hundreds of FOVs, we strongly recommend
reading the MERFISH output on a server with a large number of CPU cores.
Alternatively, some but not all MERFISH datasets store cell segmentation
in a <code>parquet</code> file, which can be more easily read into R.
This requires the installation of <a href="https://arrow.apache.org/" class="external-link"><code>arrow</code></a>.</p>
<p>The <code>SFEData</code> package (version 1.6.0 and later) provides
smaller subsets of data in the standard output format for MERFISH,
Xenium, and CosMX for testing and example purposes via the
<code>*Output</code> functions.</p>
<p>Here we read a toy dataset which is the first FOV from a real
dataset; <em>note that the first time you run this code, the code chink
in the notebook appears to stall and you need to go to the R console to
type Yes to download the dataset</em>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">dir_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/VizgenOutput.html" class="external-link">VizgenOutput</a></span><span class="op">(</span>file_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fp</span>, <span class="st">"vizgen"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; The downloaded files are in /tmp/RtmpOVsuEd/vizgen/vizgen_cellbound</span></span>
<span><span class="fu"><a href="https://fs.r-lib.org/reference/dir_tree.html" class="external-link">dir_tree</a></span><span class="op">(</span><span class="va">dir_use</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB; font-weight: bold;">/tmp/RtmpOVsuEd/vizgen/vizgen_cellbound</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #0000BB; font-weight: bold;">cell_boundaries</span></span></span>
<span><span class="co">#&gt; │   ├── feature_data_z2_z3_1.hdf5</span></span>
<span><span class="co">#&gt; │   ├── feature_data_z2_z3_2.hdf5</span></span>
<span><span class="co">#&gt; │   ├── feature_data_z2_z3_3.hdf5</span></span>
<span><span class="co">#&gt; │   └── feature_data_z2_z3_4.hdf5</span></span>
<span><span class="co">#&gt; ├── cell_boundaries.parquet</span></span>
<span><span class="co">#&gt; ├── cell_by_gene.csv</span></span>
<span><span class="co">#&gt; ├── cell_metadata.csv</span></span>
<span><span class="co">#&gt; ├── detected_transcripts.csv</span></span>
<span><span class="co">#&gt; └── <span style="color: #0000BB; font-weight: bold;">images</span></span></span>
<span><span class="co">#&gt;     ├── manifest.json</span></span>
<span><span class="co">#&gt;     ├── <span style="color: #BB00BB; font-weight: bold;">mosaic_Cellbound1_z3.tif</span></span></span>
<span><span class="co">#&gt;     ├── <span style="color: #BB00BB; font-weight: bold;">mosaic_Cellbound2_z1.tif</span></span></span>
<span><span class="co">#&gt;     ├── <span style="color: #BB00BB; font-weight: bold;">mosaic_Cellbound2_z3.tif</span></span></span>
<span><span class="co">#&gt;     ├── <span style="color: #BB00BB; font-weight: bold;">mosaic_Cellbound3_z3.tif</span></span></span>
<span><span class="co">#&gt;     ├── <span style="color: #BB00BB; font-weight: bold;">mosaic_DAPI_z1.tif</span></span></span>
<span><span class="co">#&gt;     ├── <span style="color: #BB00BB; font-weight: bold;">mosaic_DAPI_z2.tif</span></span></span>
<span><span class="co">#&gt;     ├── <span style="color: #BB00BB; font-weight: bold;">mosaic_DAPI_z3.tif</span></span></span>
<span><span class="co">#&gt;     ├── <span style="color: #BB00BB; font-weight: bold;">mosaic_PolyT_z2.tif</span></span></span>
<span><span class="co">#&gt;     └── <span style="color: #BB00BB; font-weight: bold;">mosaic_PolyT_z3.tif</span></span></span></code></pre></div>
<p>Cell segmentations from CellPose are in the
<code>cell_boundaries.parquet</code> file, which is already GeoParquet,
which means it already has the Simple Features rather than just the
coordinates of the vertices. No conversion or reformatting is needed.
The optional <code>add_molecules</code> argument can be set to
<code>TRUE</code> to read in the transcript spots; behind the scene,
this calls the <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/formatTxSpots.html" class="external-link">formatTxSpots()</a></code> function, which can also be
called separately.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe_mer</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readVizgen.html" class="external-link">readVizgen</a></span><span class="op">(</span><span class="va">dir_use</span>, z <span class="op">=</span> <span class="fl">3L</span>, add_molecules <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; 1 `.parquet` files exist:</span></span>
<span><span class="co">#&gt; /tmp/RtmpOVsuEd/vizgen/vizgen_cellbound/cell_boundaries.parquet</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; using -&gt; /tmp/RtmpOVsuEd/vizgen/vizgen_cellbound/cell_boundaries.parquet</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Cell segmentations are found in `.parquet` file</span></span>
<span><span class="co">#&gt; Removing 35 cells with area less than 15</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; filtering geometries to match 1023 cells with counts &gt; 0</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Checking polygon validity</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading transcript coordinates</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Converting transcript spots to geometry</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Writing reformatted transcript spots to disk</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 88 1023 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(88): CD4 TLL1 ... Blank-38 Blank-39</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(1023): 112824700230101267 112824700230101269 ...</span></span>
<span><span class="co">#&gt;   112824700330100848 112824700330100920</span></span>
<span><span class="co">#&gt; colData names(11): fov volume ... solidity sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : center_x center_y</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: micron</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT), cellSeg (POLYGON) </span></span>
<span><span class="co">#&gt; rowGeometries: txSpots (MULTIPOINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
<p>The unit is always in microns. To make it easier and faster to read
the data next time, the transcript spots are written to a file
<code>detected_transcripts.parquet</code> in the same directory where
the data resides:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://fs.r-lib.org/reference/dir_tree.html" class="external-link">dir_tree</a></span><span class="op">(</span><span class="va">dir_use</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB; font-weight: bold;">/tmp/RtmpOVsuEd/vizgen/vizgen_cellbound</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #0000BB; font-weight: bold;">cell_boundaries</span></span></span>
<span><span class="co">#&gt; │   ├── feature_data_z2_z3_1.hdf5</span></span>
<span><span class="co">#&gt; │   ├── feature_data_z2_z3_2.hdf5</span></span>
<span><span class="co">#&gt; │   ├── feature_data_z2_z3_3.hdf5</span></span>
<span><span class="co">#&gt; │   └── feature_data_z2_z3_4.hdf5</span></span>
<span><span class="co">#&gt; ├── cell_boundaries.parquet</span></span>
<span><span class="co">#&gt; ├── cell_by_gene.csv</span></span>
<span><span class="co">#&gt; ├── cell_metadata.csv</span></span>
<span><span class="co">#&gt; ├── detected_transcripts.csv</span></span>
<span><span class="co">#&gt; ├── detected_transcripts.parquet</span></span>
<span><span class="co">#&gt; └── <span style="color: #0000BB; font-weight: bold;">images</span></span></span>
<span><span class="co">#&gt;     ├── manifest.json</span></span>
<span><span class="co">#&gt;     ├── <span style="color: #BB00BB; font-weight: bold;">mosaic_Cellbound1_z3.tif</span></span></span>
<span><span class="co">#&gt;     ├── <span style="color: #BB00BB; font-weight: bold;">mosaic_Cellbound2_z1.tif</span></span></span>
<span><span class="co">#&gt;     ├── <span style="color: #BB00BB; font-weight: bold;">mosaic_Cellbound2_z3.tif</span></span></span>
<span><span class="co">#&gt;     ├── <span style="color: #BB00BB; font-weight: bold;">mosaic_Cellbound3_z3.tif</span></span></span>
<span><span class="co">#&gt;     ├── <span style="color: #BB00BB; font-weight: bold;">mosaic_DAPI_z1.tif</span></span></span>
<span><span class="co">#&gt;     ├── <span style="color: #BB00BB; font-weight: bold;">mosaic_DAPI_z2.tif</span></span></span>
<span><span class="co">#&gt;     ├── <span style="color: #BB00BB; font-weight: bold;">mosaic_DAPI_z3.tif</span></span></span>
<span><span class="co">#&gt;     ├── <span style="color: #BB00BB; font-weight: bold;">mosaic_PolyT_z2.tif</span></span></span>
<span><span class="co">#&gt;     └── <span style="color: #BB00BB; font-weight: bold;">mosaic_PolyT_z3.tif</span></span></span></code></pre></div>
<p>The GeoParquet file is much smaller than the original CSV file; <a href="https://parquet.apache.org/docs/overview/" class="external-link"><code>parquet</code></a>
from Apache Arrow is designed to be more memory efficient and faster to
read than CSV files and is supported in many different programming
languages, facilitating interoperability:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://fs.r-lib.org/reference/file_info.html" class="external-link">file_size</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir_use</span>, <span class="st">"detected_transcripts.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; 828K</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://fs.r-lib.org/reference/file_info.html" class="external-link">file_size</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir_use</span>, <span class="st">"detected_transcripts.parquet"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; 27.7K</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="x-xenium-output">10X Xenium output<a class="anchor" aria-label="anchor" href="#x-xenium-output"></a>
</h3>
<p>SFE supports reading the output from Xenium Onboarding Analysis (XOA)
v1 and v2 with the function <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readXenium.html" class="external-link">readXenium()</a></code>. Especially for
XOA v2, <code>arrow</code> is strongly recommended. The cell and nuclei
polygon vertices and transcript spot coordinates are in
<code>parquet</code> files (not GeoParquet). <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readXenium.html" class="external-link">readXenium()</a></code>
makes <code>sf</code> data frames from the vertices and transcript spots
and saves them as GeoParquet files.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dir_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/XeniumOutput.html" class="external-link">XeniumOutput</a></span><span class="op">(</span><span class="st">"v2"</span>, file_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fp</span>, <span class="st">"xenium"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; The downloaded files are in /tmp/RtmpOVsuEd/xenium/xenium2</span></span>
<span><span class="fu"><a href="https://fs.r-lib.org/reference/dir_tree.html" class="external-link">dir_tree</a></span><span class="op">(</span><span class="va">dir_use</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB; font-weight: bold;">/tmp/RtmpOVsuEd/xenium/xenium2</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #BB0000; font-weight: bold;">cell_boundaries.csv.gz</span></span></span>
<span><span class="co">#&gt; ├── cell_boundaries.parquet</span></span>
<span><span class="co">#&gt; ├── cell_feature_matrix.h5</span></span>
<span><span class="co">#&gt; ├── <span style="color: #BB0000; font-weight: bold;">cells.csv.gz</span></span></span>
<span><span class="co">#&gt; ├── cells.parquet</span></span>
<span><span class="co">#&gt; ├── experiment.xenium</span></span>
<span><span class="co">#&gt; ├── <span style="color: #0000BB; font-weight: bold;">morphology_focus</span></span></span>
<span><span class="co">#&gt; │   ├── <span style="color: #BB00BB; font-weight: bold;">morphology_focus_0000.ome.tif</span></span></span>
<span><span class="co">#&gt; │   ├── <span style="color: #BB00BB; font-weight: bold;">morphology_focus_0001.ome.tif</span></span></span>
<span><span class="co">#&gt; │   ├── <span style="color: #BB00BB; font-weight: bold;">morphology_focus_0002.ome.tif</span></span></span>
<span><span class="co">#&gt; │   └── <span style="color: #BB00BB; font-weight: bold;">morphology_focus_0003.ome.tif</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #BB0000; font-weight: bold;">nucleus_boundaries.csv.gz</span></span></span>
<span><span class="co">#&gt; ├── nucleus_boundaries.parquet</span></span>
<span><span class="co">#&gt; ├── <span style="color: #BB0000; font-weight: bold;">transcripts.csv.gz</span></span></span>
<span><span class="co">#&gt; └── transcripts.parquet</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># RBioFormats issue: https://github.com/aoles/RBioFormats/issues/42</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/try.html" class="external-link">try</a></span><span class="op">(</span><span class="va">sfe_xen</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readXenium.html" class="external-link">readXenium</a></span><span class="op">(</span><span class="va">dir_use</span>, add_molecules <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Must use gene symbols as row names when adding transcript spots.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Cell segmentations are found in `.parquet` file(s)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading cell and nucleus segmentations</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Making MULTIPOLYGON nuclei geometries</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Making POLYGON cell geometries</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Checking polygon validity</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Saving geometries to parquet files</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading cell metadata -&gt; `cells.parquet`</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading h5 gene count matrix</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; filtering cellSeg geometries to match 6272 cells with counts &gt; 0</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; filtering nucSeg geometries to match 6158 cells with counts &gt; 0</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading transcript coordinates</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Converting transcript spots to geometry</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Writing reformatted transcript spots to disk</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Total of 116 features/genes with no transcript detected or `min_phred` &lt; 20 are removed from SFE object</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; To keep all features -&gt; set `min_phred = NULL`</span></span>
<span><span class="op">(</span><span class="va">sfe_xen</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readXenium.html" class="external-link">readXenium</a></span><span class="op">(</span><span class="va">dir_use</span>, add_molecules <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Must use gene symbols as row names when adding transcript spots.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Preprocessed sf segmentations found</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading cell and nucleus segmentations</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading cell metadata -&gt; `cells.parquet`</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading h5 gene count matrix</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; filtering cellSeg geometries to match 6272 cells with counts &gt; 0</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; filtering nucSeg geometries to match 6158 cells with counts &gt; 0</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading transcript coordinates</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Total of 116 features/genes with no transcript detected or `min_phred` &lt; 20 are removed from SFE object</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; To keep all features -&gt; set `min_phred = NULL`</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 398 6272 </span></span>
<span><span class="co">#&gt; metadata(1): Samples</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(398): ABCC11 ACE2 ... UnassignedCodeword_0488</span></span>
<span><span class="co">#&gt;   UnassignedCodeword_0497</span></span>
<span><span class="co">#&gt; rowData names(3): ID Symbol Type</span></span>
<span><span class="co">#&gt; colnames(6272): abclkehb-1 abcnopgp-1 ... odmgoega-1 odmgojlc-1</span></span>
<span><span class="co">#&gt; colData names(9): transcript_counts control_probe_counts ...</span></span>
<span><span class="co">#&gt;   nucleus_area sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x_centroid y_centroid</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: micron</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT), cellSeg (POLYGON), nucSeg (MULTIPOLYGON) </span></span>
<span><span class="co">#&gt; rowGeometries: txSpots (MULTIPOINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
<p>The processed cell segmentation is written to
<code>cell_boundaries_sf.parquet</code> as GeoParquet for faster reading
next time, nucleus segmentation to
<code>nucleus_boundaries_sf.parquet</code>, and transcript spots to
<code>tx_spots.parquet</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://fs.r-lib.org/reference/dir_tree.html" class="external-link">dir_tree</a></span><span class="op">(</span><span class="va">dir_use</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB; font-weight: bold;">/tmp/RtmpOVsuEd/xenium/xenium2</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #BB0000; font-weight: bold;">cell_boundaries.csv.gz</span></span></span>
<span><span class="co">#&gt; ├── cell_boundaries.parquet</span></span>
<span><span class="co">#&gt; ├── cell_boundaries_sf.parquet</span></span>
<span><span class="co">#&gt; ├── cell_feature_matrix.h5</span></span>
<span><span class="co">#&gt; ├── <span style="color: #BB0000; font-weight: bold;">cells.csv.gz</span></span></span>
<span><span class="co">#&gt; ├── cells.parquet</span></span>
<span><span class="co">#&gt; ├── experiment.xenium</span></span>
<span><span class="co">#&gt; ├── <span style="color: #0000BB; font-weight: bold;">morphology_focus</span></span></span>
<span><span class="co">#&gt; │   ├── <span style="color: #BB00BB; font-weight: bold;">morphology_focus_0000.ome.tif</span></span></span>
<span><span class="co">#&gt; │   ├── <span style="color: #BB00BB; font-weight: bold;">morphology_focus_0001.ome.tif</span></span></span>
<span><span class="co">#&gt; │   ├── <span style="color: #BB00BB; font-weight: bold;">morphology_focus_0002.ome.tif</span></span></span>
<span><span class="co">#&gt; │   └── <span style="color: #BB00BB; font-weight: bold;">morphology_focus_0003.ome.tif</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #BB0000; font-weight: bold;">nucleus_boundaries.csv.gz</span></span></span>
<span><span class="co">#&gt; ├── nucleus_boundaries.parquet</span></span>
<span><span class="co">#&gt; ├── nucleus_boundaries_sf.parquet</span></span>
<span><span class="co">#&gt; ├── <span style="color: #BB0000; font-weight: bold;">transcripts.csv.gz</span></span></span>
<span><span class="co">#&gt; ├── transcripts.parquet</span></span>
<span><span class="co">#&gt; └── tx_spots.parquet</span></span></code></pre></div>
<p>The upcoming Xenium Prime (XOA v3) with 5k gene panel poses a
challenge to the <code>readXenium</code> function when making the
GeoParquet file. Writing the GeoParquet file may fail due to an error
that might be related to <a href="https://lists.apache.org/thread/9ffnb5wh9fyo614oyhtq2mo32ykn23sw" class="external-link">this</a>
and it’s not specific to the R interface to <code>arrow</code>. For the
5k panel and datasets with larger numbers of cells, I may write multiple
GeoParquet files and then concatenate them with with DuckDB for the next
release.</p>
<p>In addition, for large GeoParquet transcript spot files, with the
GDAL Parquet driver (requires GDAL version 3.5.0 or higher), we can
selectively load transcript spots of genes of interest, for
e.g. visualization purposes, with the <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readSelectTx.html" class="external-link">readSelectTx()</a></code>
function. This is the beginning of an on disk shadow of the SFE object.
<a href="https://github.com/ArtifactDB" class="external-link">Genentech’s ArtifactDB
project</a> aims to implement language-agnostic on-disk serialization of
objects such as matrices, <code>SingleCellExperiment</code>, and
<code>SpatialExperiment</code> to facilitate interoperability between
programming languages. In R, such serialization is performed with the
<code>alabaster</code> series of packages on Bioconductor. These
GeoParquet files can be the beginning of <code>alabaster.sfe</code>
whose non-geometric aspects are handled by the existing <a href="https://github.com/ArtifactDB/alabaster.sce" class="external-link"><code>alabaster.sce</code></a>
and <a href="https://github.com/ArtifactDB/alabaster.spatial" class="external-link"><code>alabaster.spatial</code></a>.</p>
</div>
<div class="section level3">
<h3 id="other-technologies">Other technologies<a class="anchor" aria-label="anchor" href="#other-technologies"></a>
</h3>
<p>Nanostring CosMX data, including transcript spots, is processed and
read in a similar manner as MERFISH and Xenium data, with the
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readCosMX.html" class="external-link">readCosMX()</a></code> function. A read function for Visium HD is in
progress. Contribution for Akoya, Molecular Cartography, and Curio
Seeker are welcome. See the <a href="https://github.com/pachterlab/SpatialFeatureExperiment/issues" class="external-link">issues</a>.</p>
</div>
<div class="section level3">
<h3 id="coercion-from-spatialexperiment">Coercion from <code>SpatialExperiment</code><a class="anchor" aria-label="anchor" href="#coercion-from-spatialexperiment"></a>
</h3>
<p>Some existing spatial -omics data analysis packages use SPE. SPE
objects can be coerced into SFE objects. If column geometries or spot
diameter are not specified, then a column geometry called “centroids”
will be created.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/read10xVisium.html" class="external-link">read10xVisium</a></span><span class="op">(</span><span class="va">samples</span>, <span class="va">sample_ids</span>, type <span class="op">=</span> <span class="st">"sparse"</span>, data <span class="op">=</span> <span class="st">"filtered"</span>, </span>
<span>  images <span class="op">=</span> <span class="st">"hires"</span>, load <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>For the coercion, column names must not be duplicate.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/make.unique.html" class="external-link">make.unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"-"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/SpatialFeatureExperiment-coercion.html" class="external-link">toSpatialFeatureExperiment</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 5 25 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(5): ENSG00000014257 ENSG00000142515 ENSG00000263639</span></span>
<span><span class="co">#&gt;   ENSG00000163810 ENSG00000149591</span></span>
<span><span class="co">#&gt; rowData names(1): symbol</span></span>
<span><span class="co">#&gt; colnames(25): GTGGCGTGCACCAGAG-1 GGTCCCATAACATAGA-1 ...</span></span>
<span><span class="co">#&gt;   TGCAATTTGGGCACGG-1 ATGCCAATCGCTCTGC-1</span></span>
<span><span class="co">#&gt; colData names(4): in_tissue array_row array_col sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : pxl_col_in_fullres pxl_row_in_fullres</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit:</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01: </span></span>
<span><span class="co">#&gt; sample02:</span></span></code></pre></div>
<p>If images are present in the SPE object, they will be converted into
<code>SpatRaster</code> when the SPE object is converted into SFE.
Plotting functions in the <code>Voyager</code> package require image
classes implemented in SFE (covered later in this workshop) to plot the
image behind the geometries because these classes in SFE have the
spatial extent needed to make sure that the images are aligned to the
geometries in the same coordinate system.</p>
</div>
<div class="section level3">
<h3 id="coercion-from-seurat">Coercion from <code>Seurat</code><a class="anchor" aria-label="anchor" href="#coercion-from-seurat"></a>
</h3>
<p>Seurat objects can be coerced into SFE objects though coercion from
SFE to Seurat is not yet implemented.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dir_extdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"SpatialFeatureExperiment"</span><span class="op">)</span></span>
<span><span class="va">obj_vis</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/serialize.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir_extdata</span>, <span class="st">"seu_vis_toy.rds"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_conv_vis</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/SpatialFeatureExperiment-coercion.html" class="external-link">toSpatialFeatureExperiment</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">obj_vis</span>, </span>
<span>                             image_scalefactors <span class="op">=</span> <span class="st">"lowres"</span>,</span>
<span>                             unit <span class="op">=</span> <span class="st">"micron"</span>,</span>
<span>                             BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span><span class="op">)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Seurat Assays found: RNA</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; RNA -&gt; will be used as 'Main Experiment'</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Seurat spatial object found: VisiumV1</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; 'full_res_image_pixel' units will be used -&gt;</span></span>
<span><span class="co">#&gt; ie 'imagerow' &amp; 'imagecol' without scaling factors</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; set `unit = 'micron'` to convert spot coordinates to micron space</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Generating `sf` geometries</span></span>
<span><span class="co">#&gt; Warning: Layer 'data' is empty</span></span>
<span><span class="co">#&gt; Warning: Layer 'scale.data' is empty</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Creating `SFE` object -&gt; sample01</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Converting pixels to microns</span></span>
<span><span class="va">sfe_conv_vis</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 5 12 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(5): ACPP KLK3 MSMB TGM4 TAGLN</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(12): GTGGCGTGCACCAGAG-1 GGTCCCATAACATAGA-1 ...</span></span>
<span><span class="co">#&gt;   CTTCCTGCATATTTAC-1 CAATATGTAGATTTAC-1</span></span>
<span><span class="co">#&gt; colData names(7): orig.ident nCount_RNA ... in_tissue sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: RNA</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : X Y</span></span>
<span><span class="co">#&gt; imgData names(0):</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: micron</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="getting-setting-and-plotting-fields-of-sfe-objects">Getting, setting, and plotting fields of SFE objects<a class="anchor" aria-label="anchor" href="#getting-setting-and-plotting-fields-of-sfe-objects"></a>
</h2>
<p>Here we demonstrate operations on SFE objects on the Xenium and
MERFISH datasets read in just now, introducing the getters, setters, and
plotting functions along the way with the flow of a basic exploratory
analysis.</p>
<div class="section level3">
<h3 id="singlecellexperiment-getters-and-setters">
<code>SingleCellExperiment</code> getters and setters<a class="anchor" aria-label="anchor" href="#singlecellexperiment-getters-and-setters"></a>
</h3>
<p>Each SFE object is an SCE object as SFE builds on top of SCE, or
inherits from SCE, so all the SCE methods apply. Here “inherits” is just
like each bioinformatician is a human, where “bioinformatician” is a bit
like SFE and “human” is like SCE. Here we go over SCE getters and
setters.</p>
<p>At the center of SCE is the gene count matrix. You can get or set the
gene count matrix with <code>counts</code> function:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe_xen</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span>
<span><span class="co">#&gt; &lt;6 x 6272&gt; sparse DelayedMatrix object of type "double":</span></span>
<span><span class="co">#&gt;        abclkehb-1 abcnopgp-1 abcobdon-1 ... odmgoega-1 odmgojlc-1</span></span>
<span><span class="co">#&gt; ABCC11          0          0          0   .          0          0</span></span>
<span><span class="co">#&gt; ACE2            1          1          1   .          0          0</span></span>
<span><span class="co">#&gt; ACKR1           0          0          0   .          0          0</span></span>
<span><span class="co">#&gt; ACTA2           0          0          0   .          0          0</span></span>
<span><span class="co">#&gt; ACTG2           0          0          0   .          0          0</span></span>
<span><span class="co">#&gt; ADAM28          1          0          1   .          0          0</span></span></code></pre></div>
<p>Here the gene count matrix was read from an HDF5 file as
<code>DelayedMatrix</code> and not fully loaded into memory; this helps
with analyzing <a href="https://bioconductor.org/packages/release//data/experiment/vignettes/TENxBrainData/inst/doc/TENxBrainData.html" class="external-link">data
that doesn’t fit into memory</a>. See <a href="https://bioconductor.org/packages/release/bioc/html/DelayedArray.html" class="external-link">the
<code>DelayedArray</code> package</a> for more info and operations on
<code>DelayedMatrix</code>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Setter</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe_xen</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">m</span></span></code></pre></div>
<p>After log normalizing data, similarly the <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts()</a></code>
function can be used to get or set the log normalized gene count
matrix.</p>
<p>The gene count matrix has metadata about the cells and genes. Use the
<code>colData</code> function to get cell metadata and
<code>rowData</code> to get gene metadata</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe_xen</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 6272 rows and 9 columns</span></span>
<span><span class="co">#&gt;            transcript_counts control_probe_counts control_codeword_counts</span></span>
<span><span class="co">#&gt;                    &lt;integer&gt;            &lt;integer&gt;               &lt;integer&gt;</span></span>
<span><span class="co">#&gt; abclkehb-1                26                    0                       0</span></span>
<span><span class="co">#&gt; abcnopgp-1                35                    0                       0</span></span>
<span><span class="co">#&gt; abcobdon-1                39                    0                       0</span></span>
<span><span class="co">#&gt; abcohgbl-1                45                    0                       0</span></span>
<span><span class="co">#&gt; abcoochm-1                61                    0                       0</span></span>
<span><span class="co">#&gt; ...                      ...                  ...                     ...</span></span>
<span><span class="co">#&gt; odmfjplg-1                12                    0                       0</span></span>
<span><span class="co">#&gt; odmfpjop-1                 3                    0                       0</span></span>
<span><span class="co">#&gt; odmglico-1                40                    0                       0</span></span>
<span><span class="co">#&gt; odmgoega-1                25                    0                       0</span></span>
<span><span class="co">#&gt; odmgojlc-1                13                    0                       0</span></span>
<span><span class="co">#&gt;            unassigned_codeword_counts deprecated_codeword_counts total_counts</span></span>
<span><span class="co">#&gt;                             &lt;integer&gt;                  &lt;integer&gt;    &lt;integer&gt;</span></span>
<span><span class="co">#&gt; abclkehb-1                          0                          0           26</span></span>
<span><span class="co">#&gt; abcnopgp-1                          0                          0           35</span></span>
<span><span class="co">#&gt; abcobdon-1                          0                          0           39</span></span>
<span><span class="co">#&gt; abcohgbl-1                          0                          0           45</span></span>
<span><span class="co">#&gt; abcoochm-1                          0                          0           61</span></span>
<span><span class="co">#&gt; ...                               ...                        ...          ...</span></span>
<span><span class="co">#&gt; odmfjplg-1                          0                          0           12</span></span>
<span><span class="co">#&gt; odmfpjop-1                          0                          0            3</span></span>
<span><span class="co">#&gt; odmglico-1                          0                          0           40</span></span>
<span><span class="co">#&gt; odmgoega-1                          0                          0           25</span></span>
<span><span class="co">#&gt; odmgojlc-1                          0                          0           13</span></span>
<span><span class="co">#&gt;            cell_area nucleus_area   sample_id</span></span>
<span><span class="co">#&gt;            &lt;numeric&gt;    &lt;numeric&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt; abclkehb-1   39.6923      18.9656    sample01</span></span>
<span><span class="co">#&gt; abcnopgp-1   60.7352      29.5322    sample01</span></span>
<span><span class="co">#&gt; abcobdon-1   55.4519      20.9977    sample01</span></span>
<span><span class="co">#&gt; abcohgbl-1   79.9266      24.4295    sample01</span></span>
<span><span class="co">#&gt; abcoochm-1   70.6244      34.1833    sample01</span></span>
<span><span class="co">#&gt; ...              ...          ...         ...</span></span>
<span><span class="co">#&gt; odmfjplg-1  63.03813     29.35156    sample01</span></span>
<span><span class="co">#&gt; odmfpjop-1   9.16672      9.16672    sample01</span></span>
<span><span class="co">#&gt; odmglico-1  82.54563     16.07563    sample01</span></span>
<span><span class="co">#&gt; odmgoega-1 112.84547      5.73484    sample01</span></span>
<span><span class="co">#&gt; odmgojlc-1  47.00766     12.19219    sample01</span></span></code></pre></div>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_xen</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 398 rows and 3 columns</span></span>
<span><span class="co">#&gt;                                             ID                 Symbol</span></span>
<span><span class="co">#&gt;                                    &lt;character&gt;            &lt;character&gt;</span></span>
<span><span class="co">#&gt; ABCC11                         ENSG00000121270                 ABCC11</span></span>
<span><span class="co">#&gt; ACE2                           ENSG00000130234                   ACE2</span></span>
<span><span class="co">#&gt; ACKR1                          ENSG00000213088                  ACKR1</span></span>
<span><span class="co">#&gt; ACTA2                          ENSG00000107796                  ACTA2</span></span>
<span><span class="co">#&gt; ACTG2                          ENSG00000163017                  ACTG2</span></span>
<span><span class="co">#&gt; ...                                        ...                    ...</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0461 UnassignedCodeword_0.. UnassignedCodeword_0..</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0469 UnassignedCodeword_0.. UnassignedCodeword_0..</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0479 UnassignedCodeword_0.. UnassignedCodeword_0..</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0488 UnassignedCodeword_0.. UnassignedCodeword_0..</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0497 UnassignedCodeword_0.. UnassignedCodeword_0..</span></span>
<span><span class="co">#&gt;                                        Type</span></span>
<span><span class="co">#&gt;                                 &lt;character&gt;</span></span>
<span><span class="co">#&gt; ABCC11                      Gene Expression</span></span>
<span><span class="co">#&gt; ACE2                        Gene Expression</span></span>
<span><span class="co">#&gt; ACKR1                       Gene Expression</span></span>
<span><span class="co">#&gt; ACTA2                       Gene Expression</span></span>
<span><span class="co">#&gt; ACTG2                       Gene Expression</span></span>
<span><span class="co">#&gt; ...                                     ...</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0461 Unassigned Codeword</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0469 Unassigned Codeword</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0479 Unassigned Codeword</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0488 Unassigned Codeword</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0497 Unassigned Codeword</span></span></code></pre></div>
<p>Just like in Seurat, the SCE object can be subsetted like a matrix,
for example, to remove low quality cells with too few transcript counts
and genes that are not detected. <code>colData</code> columns in SCE can
be accessed with the <code>$</code> operator as if getting a column from
a data frame. When it comes to the data frame analogy, also see the <a href="https://www.nature.com/articles/s41592-024-02299-2" class="external-link">tidyOmics
packages</a> that brings the Tidyverse to -omics data and gives a
unified user interface uniting Seurat and SCE analyses.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_xen</span> <span class="op">&lt;-</span> <span class="va">sfe_xen</span><span class="op">[</span>, <span class="va">sfe_xen</span><span class="op">$</span><span class="va">total_counts</span> <span class="op">&gt;</span> <span class="fl">5</span><span class="op">]</span></span>
<span><span class="va">sfe_xen</span> <span class="op">&lt;-</span> <span class="va">sfe_xen</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe_xen</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">sfe_xen</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]  394 5749</span></span></code></pre></div>
<p>PCA is part of the standard scRNA-seq data analysis workflow. Here
we’ll first normalize the data and then perform PCA and get the PCA
results. Using cell area as size factors comes from <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-024-03303-w" class="external-link">this
paper</a>; using total counts as in scRNA-seq is inappropriate when we
have a curated gene panel and can blunt biological signals.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_xen</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sfe_xen</span>, size.factors <span class="op">=</span> <span class="va">sfe_xen</span><span class="op">$</span><span class="va">cell_area</span><span class="op">)</span></span>
<span><span class="co"># Log counts getter</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">sfe_xen</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; &lt;6 x 5749&gt; sparse DelayedMatrix object of type "double":</span></span>
<span><span class="co">#&gt;        abclkehb-1 abcnopgp-1 abcobdon-1 ... odmgoega-1 odmgojlc-1</span></span>
<span><span class="co">#&gt; ABCC11   0.000000   0.000000   0.000000   .          0          0</span></span>
<span><span class="co">#&gt; ACE2     1.346809   1.006276   1.073702   .          0          0</span></span>
<span><span class="co">#&gt; ACKR1    0.000000   0.000000   0.000000   .          0          0</span></span>
<span><span class="co">#&gt; ACTA2    0.000000   0.000000   0.000000   .          0          0</span></span>
<span><span class="co">#&gt; ACTG2    0.000000   0.000000   0.000000   .          0          0</span></span>
<span><span class="co">#&gt; ADAM28   1.346809   0.000000   1.073702   .          0          0</span></span></code></pre></div>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_xen</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sfe_xen</span>, ncomponents <span class="op">=</span> <span class="fl">20</span>,</span>
<span>                  exprs_values <span class="op">=</span> <span class="st">"logcounts"</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Later we will see that Voyager spatial analysis functions are modeled
after <code>runPCA</code> un user interface. The <code>reducedDim</code>
function can be used to get and set dimension reduction results. User
interfaces to get or set the geometries and spatial graphs emulate those
of <code>reducedDims</code> and <code>row/colPairs</code> in
<code>SingleCellExperiment</code>. Column and row geometries also
emulate <code>reducedDims</code> in internal implementation, while
annotation geometries and spatial graphs differ.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sfe_xen</span>, <span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pca_res</span><span class="op">)</span></span>
<span><span class="co">#&gt;                 PC1       PC2         PC3       PC4        PC5        PC6</span></span>
<span><span class="co">#&gt; abclkehb-1 4.104976 -5.209892 -0.32024759 -3.448198 -0.8745107 -1.4135185</span></span>
<span><span class="co">#&gt; abcnopgp-1 3.698699 -3.204301  0.37888409 -2.166183 -0.3577043 -0.6102030</span></span>
<span><span class="co">#&gt; abcobdon-1 3.684348 -3.824745 -0.26077213 -2.621562 -0.8730314 -0.2183592</span></span>
<span><span class="co">#&gt; abcohgbl-1 3.575645 -4.547060  0.01390075 -3.020987 -0.5612930 -0.9650647</span></span>
<span><span class="co">#&gt; abcoochm-1 5.867807 -4.898515  0.19077378 -2.473097 -2.1675873 -2.8156440</span></span>
<span><span class="co">#&gt; abcoplda-1 3.219883 -5.331865  0.42655776 -3.515862 -0.9702069 -1.5188901</span></span>
<span><span class="co">#&gt;                   PC7        PC8         PC9     PC10       PC11       PC12</span></span>
<span><span class="co">#&gt; abclkehb-1 -0.9713220 -1.9336509 -0.08400579 3.588110 -0.1042168 -0.6512118</span></span>
<span><span class="co">#&gt; abcnopgp-1 -0.1503387 -0.4381665 -0.23839749 2.898649 -0.8889157 -0.8333472</span></span>
<span><span class="co">#&gt; abcobdon-1 -0.7347887 -0.6500136 -0.04385059 1.688362 -0.9957585 -0.7186998</span></span>
<span><span class="co">#&gt; abcohgbl-1 -0.6458784 -1.3219623 -0.27045651 2.627548 -0.3681193 -0.6257008</span></span>
<span><span class="co">#&gt; abcoochm-1 -0.3622780 -1.9080242  0.11771026 2.126753 -0.4656962 -1.3993553</span></span>
<span><span class="co">#&gt; abcoplda-1  0.5366565 -1.2170732 -1.47320560 2.013572 -0.6720101 -0.9997679</span></span>
<span><span class="co">#&gt;                  PC13        PC14        PC15       PC16       PC17        PC18</span></span>
<span><span class="co">#&gt; abclkehb-1 -1.1038552  0.90477874 -0.13758657  1.0476007  1.2129374 -0.97232802</span></span>
<span><span class="co">#&gt; abcnopgp-1 -0.2487955  0.09489004 -0.79986738 -0.7257234  0.6517269  0.59408468</span></span>
<span><span class="co">#&gt; abcobdon-1 -0.6272968 -0.16732126 -0.64625317 -0.3599726  0.2004521 -1.14290957</span></span>
<span><span class="co">#&gt; abcohgbl-1 -0.1259936  0.15850392  1.64506087  0.8239103  0.3532861 -0.72690582</span></span>
<span><span class="co">#&gt; abcoochm-1  1.3791996 -0.12748514  0.04746375  0.7644174  0.4889616  0.09668229</span></span>
<span><span class="co">#&gt; abcoplda-1 -0.1754514 -0.24161746 -0.12717546  0.3201031 -0.2695271 -1.72548119</span></span>
<span><span class="co">#&gt;                  PC19       PC20</span></span>
<span><span class="co">#&gt; abclkehb-1 -0.3029037  1.1541254</span></span>
<span><span class="co">#&gt; abcnopgp-1  0.2247111 -0.3280252</span></span>
<span><span class="co">#&gt; abcobdon-1  0.4850882  0.8060317</span></span>
<span><span class="co">#&gt; abcohgbl-1 -0.9053308 -0.2138684</span></span>
<span><span class="co">#&gt; abcoochm-1 -0.4697169  0.8896973</span></span>
<span><span class="co">#&gt; abcoplda-1  0.4623914 -0.9390845</span></span></code></pre></div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set PCA embeddings say if you ran PCA elsewhere</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sfe_xen</span>, <span class="st">"PCA"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">pca_res</span></span></code></pre></div>
<p>Which dimension reductions are present?</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDimNames</a></span><span class="op">(</span><span class="va">sfe_xen</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "PCA"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="column-geometries">Column geometries<a class="anchor" aria-label="anchor" href="#column-geometries"></a>
</h3>
<p>Column geometries or <code>colGeometries</code> are the geometries
that correspond to columns of the gene count matrix, such as Visium
spots and cells in datasets from a single cell resolution technology.
Each SFE object can have multiple column geometries. For example, in a
dataset with single cell resolution, whole cell segmentation and nuclei
segmentation are two different <code>colGeometries</code>. However, for
Visium, the spot polygons are the only <code>colGeometry</code>
obviously relevant, though users can add other geometries such as
results of geometric operations on the spot polygons. The different
geometries can be get or set with their names; “spotPoly” is the
standard name for Visium spot polygons, “cellSeg” for cell segmentation,
and “nucSeg” for nucleus segmentation. But any name can be used; say
cell segmentations from different algorithms can be stored here under
different names, say “CellPost” and “watershed”, although different
segmentations can lead to different grene counts per cell and different
number of cells.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get Visium spot polygons</span></span>
<span><span class="op">(</span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">colGeometry</a></span><span class="op">(</span><span class="va">sfe_xen</span>, <span class="st">"cellSeg"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Simple feature collection with 5749 features and 1 field</span></span>
<span><span class="co">#&gt; Geometry type: POLYGON</span></span>
<span><span class="co">#&gt; Dimension:     XY</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: 0 ymin: -1008.1 xmax: 1026.163 ymax: 0</span></span>
<span><span class="co">#&gt; CRS:           NA</span></span>
<span><span class="co">#&gt; First 10 features:</span></span>
<span><span class="co">#&gt;            label_id                       geometry</span></span>
<span><span class="co">#&gt; abclkehb-1      703 POLYGON ((389.9376 -922.462...</span></span>
<span><span class="co">#&gt; abcnopgp-1      704 POLYGON ((378.8875 -934.787...</span></span>
<span><span class="co">#&gt; abcobdon-1      705 POLYGON ((381.4376 -931.6, ...</span></span>
<span><span class="co">#&gt; abcohgbl-1      706 POLYGON ((386.7501 -925.225...</span></span>
<span><span class="co">#&gt; abcoochm-1      707 POLYGON ((416.075 -896.9625...</span></span>
<span><span class="co">#&gt; abcoplda-1      708 POLYGON ((401.6251 -909.925...</span></span>
<span><span class="co">#&gt; abcopngk-1      709 POLYGON ((415.0125 -911.412...</span></span>
<span><span class="co">#&gt; abcphhlp-1      710 POLYGON ((407.575 -906.525,...</span></span>
<span><span class="co">#&gt; abcpmfic-1      711 POLYGON ((420.1125 -905.462...</span></span>
<span><span class="co">#&gt; abdaapec-1      712 POLYGON ((408.85 -900.3625,...</span></span></code></pre></div>
<p>Here we get a <code>sf</code> data frame, which is just like a
regular data frame but with a special <code>geometry</code> column. Now
plot these cells</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-34-1.png" width="700"></p>
<p>The <code>Voyager</code> package can plot the geometries without
other data with the <code><a href="https://rdrr.io/pkg/Voyager/man/plotGeometry.html" class="external-link">plotGeometry()</a></code> function. Note that the
arguments have changed since version 1.7.0 (<code>devel</code>
version)</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotGeometry.html" class="external-link">plotGeometry</a></span><span class="op">(</span><span class="va">sfe_xen</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, show_axes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotGeometry.html" class="external-link">plotGeometry</a></span><span class="op">(</span><span class="va">sfe_xen</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, show_axes <span class="op">=</span> <span class="cn">TRUE</span>, dark <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-35-1.png" width="700"></p>
<p>A bounding box can be used to zoom into a smaller region in
<code>plotGeometry</code> and any other geometry plotting function in
<code>Voyager</code> such as <code>plotSpatialFeature</code>.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bbox</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="fl">0</span>, xmax <span class="op">=</span> <span class="fl">200</span>, ymin <span class="op">=</span> <span class="op">-</span><span class="fl">600</span>, ymax <span class="op">=</span> <span class="op">-</span><span class="fl">400</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotGeometry.html" class="external-link">plotGeometry</a></span><span class="op">(</span><span class="va">sfe_xen</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, bbox <span class="op">=</span> <span class="va">bbox</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-37-1.png" width="700"></p>
<p>Multiple geometries, such as cells and nuclei, can be plotted at once
(only supported in version 1.7.0 or later)</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotGeometry.html" class="external-link">plotGeometry</a></span><span class="op">(</span><span class="va">sfe_xen</span>, colGeometryName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cellSeg"</span>, <span class="st">"nucSeg"</span><span class="op">)</span>, bbox <span class="op">=</span> <span class="va">bbox</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-38-1.png" width="700"></p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set colGeometry</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">colGeometry</a></span><span class="op">(</span><span class="va">sfe_xen</span>, <span class="st">"cellSeg"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">cells</span></span></code></pre></div>
<p>To see which <code>colGeometries</code> are present in the SFE
object:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">colGeometryNames</a></span><span class="op">(</span><span class="va">sfe_xen</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "centroids" "cellSeg"   "nucSeg"</span></span></code></pre></div>
<p>There are shorthands for some specific column or row geometries. For
example, <code>spotPoly(sfe)</code> is equivalent to
<code>colGeometry(sfe, "spotPoly")</code>, <code>cellSeg(sfe)</code> is
equivalent to <code>colGeometry(sfe, "cellSeg")</code>, and
<code>nucSeg(sfe)</code> to <code>colGeometry(sfe, "nucSeg")</code>.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Getter</span></span>
<span><span class="op">(</span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">cellSeg</a></span><span class="op">(</span><span class="va">sfe_xen</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Simple feature collection with 5749 features and 1 field</span></span>
<span><span class="co">#&gt; Geometry type: POLYGON</span></span>
<span><span class="co">#&gt; Dimension:     XY</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: 0 ymin: -1008.1 xmax: 1026.163 ymax: 0</span></span>
<span><span class="co">#&gt; CRS:           NA</span></span>
<span><span class="co">#&gt; First 10 features:</span></span>
<span><span class="co">#&gt;            label_id                       geometry</span></span>
<span><span class="co">#&gt; abclkehb-1      703 POLYGON ((389.9376 -922.462...</span></span>
<span><span class="co">#&gt; abcnopgp-1      704 POLYGON ((378.8875 -934.787...</span></span>
<span><span class="co">#&gt; abcobdon-1      705 POLYGON ((381.4376 -931.6, ...</span></span>
<span><span class="co">#&gt; abcohgbl-1      706 POLYGON ((386.7501 -925.225...</span></span>
<span><span class="co">#&gt; abcoochm-1      707 POLYGON ((416.075 -896.9625...</span></span>
<span><span class="co">#&gt; abcoplda-1      708 POLYGON ((401.6251 -909.925...</span></span>
<span><span class="co">#&gt; abcopngk-1      709 POLYGON ((415.0125 -911.412...</span></span>
<span><span class="co">#&gt; abcphhlp-1      710 POLYGON ((407.575 -906.525,...</span></span>
<span><span class="co">#&gt; abcpmfic-1      711 POLYGON ((420.1125 -905.462...</span></span>
<span><span class="co">#&gt; abdaapec-1      712 POLYGON ((408.85 -900.3625,...</span></span></code></pre></div>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Setter</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">cellSeg</a></span><span class="op">(</span><span class="va">sfe_xen</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">cells</span></span></code></pre></div>
<p>Exercise: Get and plot nucleus segmentations from
<code>sfe_xen</code></p>
<p>While we have <code>DelayedArray</code> for large gene count matrices
that don’t fit into memory, at present the geometries are in memory. I
haven’t gotten into memory problems from the geometries even from
hundreds of thousands of cell polygons on a 2017 MacBook Pro with 8 GB
of RAM yet. In a future version, we may try <a href="https://sedona.apache.org/1.4.1/" class="external-link">sedona</a> or <a href="https://duckdb.org/docs/extensions/spatial.html" class="external-link">DuckDB</a>
combined with <a href="https://github.com/JosiahParry/sdf" class="external-link"><code>sdf</code></a> to be
more general than <code>sf</code> to allow geometric operations without
loading everything into memory.</p>
</div>
<div class="section level3">
<h3 id="row-geometries">Row geometries<a class="anchor" aria-label="anchor" href="#row-geometries"></a>
</h3>
<p>The <code>rowGeometry</code> getter and setter have pretty much the
same user interface as the getters and setters covered above:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">rg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/rowGeometries.html" class="external-link">rowGeometry</a></span><span class="op">(</span><span class="va">sfe_xen</span>, <span class="st">"txSpots"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Simple feature collection with 394 features and 2 fields</span></span>
<span><span class="co">#&gt; Geometry type: MULTIPOINT</span></span>
<span><span class="co">#&gt; Dimension:     XYZ</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: 0.001159668 ymin: -1008.098 xmax: 1026.159 ymax: -4.473946</span></span>
<span><span class="co">#&gt; z_range:       zmin: 13.56251 zmax: 27.21318</span></span>
<span><span class="co">#&gt; CRS:           NA</span></span>
<span><span class="co">#&gt; First 10 features:</span></span>
<span><span class="co">#&gt;            gene codeword_index                       geometry</span></span>
<span><span class="co">#&gt; ABCC11   ABCC11             87 MULTIPOINT Z ((111.2996 -47...</span></span>
<span><span class="co">#&gt; ACE2       ACE2             31 MULTIPOINT Z ((204.9438 -21...</span></span>
<span><span class="co">#&gt; ACKR1     ACKR1            349 MULTIPOINT Z ((8.244019 -88...</span></span>
<span><span class="co">#&gt; ACTA2     ACTA2            342 MULTIPOINT Z ((4.195129 -40...</span></span>
<span><span class="co">#&gt; ACTG2     ACTG2            231 MULTIPOINT Z ((6.257874 -30...</span></span>
<span><span class="co">#&gt; ADAM28   ADAM28            119 MULTIPOINT Z ((0.2313232 -4...</span></span>
<span><span class="co">#&gt; ADAMTS1 ADAMTS1            242 MULTIPOINT Z ((5.13385 -321...</span></span>
<span><span class="co">#&gt; ADGRE1   ADGRE1             24 MULTIPOINT Z ((119.8132 -7....</span></span>
<span><span class="co">#&gt; ADGRL4   ADGRL4            132 MULTIPOINT Z ((4.375183 -32...</span></span>
<span><span class="co">#&gt; ADH1C     ADH1C             92 MULTIPOINT Z ((1.244507 -59...</span></span></code></pre></div>
<p>Note the <code>MULTIPOINT Z</code> here; this is 3D. Geometric
operations in <code>sf</code> such as finding intersections do work on
3D geometries but only for the x and y dimensions, not for z, because in
geographical space, we largely live on the surface of Earth. This is
mostly fine for histological space as well because the vast majority of
spatial -omics data come from thin sections with much higher x and y
resolution much larger x and y extent than z, although there are 3D
datasets with numerous serial sections or thick sections with 3D cell
segmentations. Since in those cases, the data usually come in discrete
z-planes, we might just use multiple polygon geometry columns each for
one z-plane instead of something like a mesh in the CGI tradition.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Setter</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/rowGeometries.html" class="external-link">rowGeometry</a></span><span class="op">(</span><span class="va">sfe_xen</span>, <span class="st">"txSpots"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">rg</span></span></code></pre></div>
<p>In the case of transcript spots, there’s a special convenience
function <code>txSpots</code>; I haven’t thought of any other use for
<code>rowGeometries</code> but this is kept general, allowing any type
of geometry, in case other use cases come up.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/rowGeometries.html" class="external-link">txSpots</a></span><span class="op">(</span><span class="va">sfe_xen</span><span class="op">)</span></span>
<span><span class="co">#&gt; Simple feature collection with 394 features and 2 fields</span></span>
<span><span class="co">#&gt; Geometry type: MULTIPOINT</span></span>
<span><span class="co">#&gt; Dimension:     XYZ</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: 0.001159668 ymin: -1008.098 xmax: 1026.159 ymax: -4.473946</span></span>
<span><span class="co">#&gt; z_range:       zmin: 13.56251 zmax: 27.21318</span></span>
<span><span class="co">#&gt; CRS:           NA</span></span>
<span><span class="co">#&gt; First 10 features:</span></span>
<span><span class="co">#&gt;            gene codeword_index                       geometry</span></span>
<span><span class="co">#&gt; ABCC11   ABCC11             87 MULTIPOINT Z ((111.2996 -47...</span></span>
<span><span class="co">#&gt; ACE2       ACE2             31 MULTIPOINT Z ((204.9438 -21...</span></span>
<span><span class="co">#&gt; ACKR1     ACKR1            349 MULTIPOINT Z ((8.244019 -88...</span></span>
<span><span class="co">#&gt; ACTA2     ACTA2            342 MULTIPOINT Z ((4.195129 -40...</span></span>
<span><span class="co">#&gt; ACTG2     ACTG2            231 MULTIPOINT Z ((6.257874 -30...</span></span>
<span><span class="co">#&gt; ADAM28   ADAM28            119 MULTIPOINT Z ((0.2313232 -4...</span></span>
<span><span class="co">#&gt; ADAMTS1 ADAMTS1            242 MULTIPOINT Z ((5.13385 -321...</span></span>
<span><span class="co">#&gt; ADGRE1   ADGRE1             24 MULTIPOINT Z ((119.8132 -7....</span></span>
<span><span class="co">#&gt; ADGRL4   ADGRL4            132 MULTIPOINT Z ((4.375183 -32...</span></span>
<span><span class="co">#&gt; ADH1C     ADH1C             92 MULTIPOINT Z ((1.244507 -59...</span></span></code></pre></div>
<p>The <code><a href="https://rdrr.io/pkg/Voyager/man/plotGeometry.html" class="external-link">plotGeometry()</a></code> function can plot transcript spots,
for all genes or for a few selected genes, distinguished by point
shape.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># bbox is used due to overplotting if plotting the entire example data</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotGeometry.html" class="external-link">plotGeometry</a></span><span class="op">(</span><span class="va">sfe_xen</span>, rowGeometryName <span class="op">=</span> <span class="st">"txSpots"</span>, bbox <span class="op">=</span> <span class="va">bbox</span>, gene <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-46-1.png" width="700"></p>
<p>That’s still a lot of spots, definitely overplotting.</p>
<p>Transcript spot density can be plotted in a 2D histogram, here using
20 micron hexagonal grid to show the density of transcripts from all
genes</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotTxBin2D.html" class="external-link">plotTxBin2D</a></span><span class="op">(</span><span class="va">sfe_xen</span>, binwidth <span class="op">=</span> <span class="fl">20</span>, hex <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-47-1.png" width="700"></p>
<p>Plot a few genes here:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span><span class="va">genes_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe_xen</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_xen</span><span class="op">)</span><span class="op">$</span><span class="va">Type</span> <span class="op">==</span> <span class="st">"Gene Expression"</span><span class="op">]</span>, <span class="fl">6</span><span class="op">)</span></span>
<span><span class="co"># Genes that don't have spots in this bbox are not plotted</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotGeometry.html" class="external-link">plotGeometry</a></span><span class="op">(</span><span class="va">sfe_xen</span>, rowGeometryName <span class="op">=</span> <span class="st">"txSpots"</span>, bbox <span class="op">=</span> <span class="va">bbox</span>, gene <span class="op">=</span> <span class="va">genes_use</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-48-1.png" width="700"></p>
<p>Or plot the transcript spots with other geometries</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotGeometry.html" class="external-link">plotGeometry</a></span><span class="op">(</span><span class="va">sfe_xen</span>, colGeometryName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cellSeg"</span>, <span class="st">"nucSeg"</span><span class="op">)</span>,</span>
<span>             rowGeometryName <span class="op">=</span> <span class="st">"txSpots"</span>, bbox <span class="op">=</span> <span class="va">bbox</span>, gene <span class="op">=</span> <span class="va">genes_use</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-49-1.png" width="700"></p>
<p>Transcript spots can also be plotted along with cell level data</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_xen</span>, <span class="st">"total_counts"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, </span>
<span>                   rowGeometryName <span class="op">=</span> <span class="st">"txSpots"</span>, rowGeometryFeatures <span class="op">=</span> <span class="va">genes_use</span>,</span>
<span>                   bbox <span class="op">=</span> <span class="va">bbox</span>, tx_fixed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-50-1.png" width="700"></p>
<p>Or plot in dark theme</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_xen</span>, <span class="st">"total_counts"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, </span>
<span>                   rowGeometryName <span class="op">=</span> <span class="st">"txSpots"</span>, rowGeometryFeatures <span class="op">=</span> <span class="va">genes_use</span>,</span>
<span>                   bbox <span class="op">=</span> <span class="va">bbox</span>, dark <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                   tx_fixed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1</span>, color <span class="op">=</span> <span class="st">"lightgray"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-51-1.png" width="700"></p>
<p>When cell level data and transcript spots of the same genes are
plotted, each gene will be plotted in a separate panel</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_xen</span>, <span class="va">genes_use</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, </span>
<span>                   rowGeometryName <span class="op">=</span> <span class="st">"txSpots"</span>, rowGeometryFeatures <span class="op">=</span> <span class="va">genes_use</span>,</span>
<span>                   bbox <span class="op">=</span> <span class="va">bbox</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-52-1.png" width="700"></p>
<p>Note that some transcript spots are not in cells; spots in
<code>rowGeometires</code> don’t have to be assigned to cells. While the
transcript spots of this data subset fits into memory, when the GDAL
Parquet driver is available, the plotting functions can call
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readSelectTx.html" class="external-link">readSelectTx()</a></code> behind the scene to only selectively read
transcripts from a few genes rather than all genes.</p>
<p>Here transcript spots are <code>sf</code> data frames. Recent
versions of <code>sf</code> can convert<code>sf</code> to objects in the
<a href="https://spatstat.org" class="external-link"><code>spatstat</code></a> package for <a href="https://r-spatial.org/book/11-PointPattern.html" class="external-link">spatial point
pattern analyses</a>.</p>
</div>
<div class="section level3">
<h3 id="annotation">Annotation<a class="anchor" aria-label="anchor" href="#annotation"></a>
</h3>
<p>Annotation geometries can be get or set with
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">annotGeometry()</a></code>. In column or row geometries, the number of
rows of the <code>sf</code> data frame (i.e. the number of geometries in
the data frame) is constrained by the number of rows or columns of the
gene count matrix respectively, because just like <code>rowData</code>
and <code>colData</code>, each row of a <code>rowGeometry</code> or
<code>colGeometry</code> <code>sf</code> data frame must correspond to a
row or column of the gene count matrix respectively. In contrast, an
<code>annotGeometry</code> <code>sf</code> data frame can have any
dimension, not constrained by the dimension of the gene count
matrix.</p>
<p>This Xenium dataset doesn’t come with <code>annotGeometries</code>,
but I used QuPath to annotate some just for the purpose of demonstration
and testing. While it looks like two sides of a river in this section,
it’s actually not two separate pieces of tissue because the original
tissue is 3 dimensional.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pieces</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/serialize.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/pieces.rds"</span>, package <span class="op">=</span> <span class="st">"SpatialFeatureExperiment"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pieces</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r-spatial.github.io/sf/reference/sf.html" class="external-link">st_sf</a></span><span class="op">(</span>geometry <span class="op">=</span> <span class="va">pieces</span>, sample_id <span class="op">=</span> <span class="st">"sample01"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">annotGeometry</a></span><span class="op">(</span><span class="va">sfe_xen</span>, <span class="st">"pieces"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">pieces</span></span></code></pre></div>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Getter, by name or index</span></span>
<span><span class="op">(</span><span class="va">tb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">annotGeometry</a></span><span class="op">(</span><span class="va">sfe_xen</span>, <span class="st">"pieces"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Simple feature collection with 2 features and 1 field</span></span>
<span><span class="co">#&gt; Geometry type: POLYGON</span></span>
<span><span class="co">#&gt; Dimension:     XY</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: 0 ymin: -1008.1 xmax: 1025.95 ymax: 0</span></span>
<span><span class="co">#&gt; CRS:           NA</span></span>
<span><span class="co">#&gt;   sample_id                       geometry</span></span>
<span><span class="co">#&gt; 1  sample01 POLYGON ((991.1 -252.45, 95...</span></span>
<span><span class="co">#&gt; 2  sample01 POLYGON ((0 0, 0 -1008.1, 2...</span></span></code></pre></div>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">tb</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-55-1.png" width="700"></p>
<p>Or use <code>plotGeometry</code></p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotGeometry.html" class="external-link">plotGeometry</a></span><span class="op">(</span><span class="va">sfe_xen</span>, annotGeometryName <span class="op">=</span> <span class="st">"pieces"</span>, show_axes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-56-1.png" width="700"></p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotGeometry.html" class="external-link">plotGeometry</a></span><span class="op">(</span><span class="va">sfe_xen</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, annotGeometryName <span class="op">=</span> <span class="st">"pieces"</span>,</span>
<span>             show_axes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-57-1.png" width="700"></p>
<p>See which <code>annotGeometries</code> are present in the SFE
object:</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">annotGeometryNames</a></span><span class="op">(</span><span class="va">sfe_xen</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "pieces"</span></span></code></pre></div>
<p>There are shorthands for specific annotation geometries. For example,
<code>tissueBoundary(sfe)</code> is equivalent to
<code>annotGeometry(sfe, "tissueBoundary")</code>.
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">cellSeg()</a></code> (cell segmentation) and <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">nucSeg()</a></code>
(nuclei segmentation) would first query <code>colGeometries</code> (for
single cell, single molecule technologies, equivalent to
<code>colGeometry(sfe, "cellSeg")</code> or
<code>colGeometry(sfe, "nucSeg")</code>), and if not found, they will
query <code>annotGeometries</code> (for array capture and
microdissection technologies, equivalent to
<code>annotGeometry(sfe, "cellSeg")</code> or
<code>annotGeometry(sfe, "nucSeg")</code>).</p>
</div>
<div class="section level3">
<h3 id="spatial-graphs">Spatial graphs<a class="anchor" aria-label="anchor" href="#spatial-graphs"></a>
</h3>
<p>The spatial neighborhood graphs for Visium spots are stored in the
<code>colGraphs</code> field, which has similar user interface as
<code>colGeometries</code>. SFE also wraps all methods to find the
spatial neighborhood graph implemented in the <code>spdep</code>
package, and triangulation is used here as demonstration.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findSpatialNeighbors-SpatialFeatureExperiment-method.html" class="external-link">findSpatialNeighbors</a></span><span class="op">(</span><span class="va">sfe_xen</span>, MARGIN <span class="op">=</span> <span class="fl">2</span>, method <span class="op">=</span> <span class="st">"tri2nb"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Characteristics of weights list object:</span></span>
<span><span class="co">#&gt; Neighbour list object:</span></span>
<span><span class="co">#&gt; Number of regions: 5749 </span></span>
<span><span class="co">#&gt; Number of nonzero links: 34444 </span></span>
<span><span class="co">#&gt; Percentage nonzero weights: 0.1042147 </span></span>
<span><span class="co">#&gt; Average number of links: 5.991303 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Weights style: W </span></span>
<span><span class="co">#&gt; Weights constants summary:</span></span>
<span><span class="co">#&gt;      n       nn   S0      S1       S2</span></span>
<span><span class="co">#&gt; W 5749 33051001 5749 1955.66 23350.69</span></span></code></pre></div>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set graph by name</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_xen</span>, <span class="st">"triangulation"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">g</span></span></code></pre></div>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get graph by name</span></span>
<span><span class="op">(</span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_xen</span>, <span class="st">"triangulation"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Characteristics of weights list object:</span></span>
<span><span class="co">#&gt; Neighbour list object:</span></span>
<span><span class="co">#&gt; Number of regions: 5749 </span></span>
<span><span class="co">#&gt; Number of nonzero links: 34444 </span></span>
<span><span class="co">#&gt; Percentage nonzero weights: 0.1042147 </span></span>
<span><span class="co">#&gt; Average number of links: 5.991303 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Weights style: W </span></span>
<span><span class="co">#&gt; Weights constants summary:</span></span>
<span><span class="co">#&gt;      n       nn   S0      S1       S2</span></span>
<span><span class="co">#&gt; W 5749 33051001 5749 1955.66 23350.69</span></span></code></pre></div>
<p>Plot the graph in space</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotColGraph.html" class="external-link">plotColGraph</a></span><span class="op">(</span><span class="va">sfe_xen</span>, colGraphName <span class="op">=</span> <span class="st">"triangulation"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-62-1.png" width="700"></p>
<p>Alternatively use k nearest neighbor graph</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_xen</span>, <span class="st">"knn"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findSpatialNeighbors-SpatialFeatureExperiment-method.html" class="external-link">findSpatialNeighbors</a></span><span class="op">(</span><span class="va">sfe_xen</span>, method <span class="op">=</span> <span class="st">"knearneigh"</span>,</span>
<span>                                                 k <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotColGraph.html" class="external-link">plotColGraph</a></span><span class="op">(</span><span class="va">sfe_xen</span>, colGraphName <span class="op">=</span> <span class="st">"knn"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-64-1.png" width="700"></p>
<p>Which graphs are present in this SFE object?</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraphNames</a></span><span class="op">(</span><span class="va">sfe_xen</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "triangulation" "knn"</span></span></code></pre></div>
<p>It would be interesting to see the implications of the choice of
spatial neighborhood graph and its edge weights to further spatial
analyses.</p>
</div>
</div>
<div class="section level2">
<h2 id="images">Images<a class="anchor" aria-label="anchor" href="#images"></a>
</h2>
<p>In SPE, the images are only used for visualization, but SFE extended
the SPE image functionality so large images don’t have to be loaded into
memory unless necessary. SFE is extensible; more image classes can be
implemented by inheriting from the virtual class
<code>AlignedSpatialImage</code>; “aligned” because a spatial extent
must be specified to align the image to the geometries, just like the
satellite image aligned to vector geometries of roads and city
boundaries. At present, the SFE package implements 3 types of images:
<code>SpatRasterImage</code>, <code>BioFormatsImage</code>, and
<code>ExtImage</code>.</p>
<div class="section level3">
<h3 id="spatrasterimage">
<code>SpatRasterImage</code><a class="anchor" aria-label="anchor" href="#spatrasterimage"></a>
</h3>
<p><code>SpatRasterImage</code> is the default, a thin wrapper around
the <a href="https://rspatial.github.io/terra/reference/SpatRaster-class.html" class="external-link"><code>SpatRaster</code></a>
class in the <code>terra</code> object to make it conform to SPE’s
requirements. Large images are not loaded into memory unless necessary
and it’s possible to only load a down sampled lower resolution version
of the image into memory. Spatial extent is part of
<code>SpatRaster</code>. The extent is important to delineate where the
image is in the coordinate system within the tissue section. This is a
more sophisticated way to make sure the image is aligned with the
geometries than the scale factor in SPE which only works for Visium and
would not allow the SPE object to be cropped.</p>
<p>In the MERFISH dataset above, the image is represented as
<code>SpatRasterImage</code>. Get images with the <code><a href="https://rdrr.io/pkg/SpatialExperiment/man/imgData-methods.html" class="external-link">getImg()</a></code>
function, and use the <code>image_id</code> function to indicate which
image to get, and if it’s left blank, the first image will be
retrieved:</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">img</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/imgData-methods.html" class="external-link">getImg</a></span><span class="op">(</span><span class="va">sfe_mer</span>, image_id <span class="op">=</span> <span class="st">"PolyT_z3"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; 695 x 695 x 1 (width x height x channels) SpatRasterImage</span></span>
<span><span class="co">#&gt; imgSource():</span></span>
<span><span class="co">#&gt;   /tmp/RtmpOVsuEd/vizgen/vizgen_cellbound/images/mosaic_PolyT_z3.tif</span></span></code></pre></div>
<p>The <code><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">imgData()</a></code> getter gets all images in the SFE object
and displays a summary including image dimension and class</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">imgData</a></span><span class="op">(</span><span class="va">sfe_mer</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 5 rows and 4 columns</span></span>
<span><span class="co">#&gt;     sample_id      image_id                          data scaleFactor</span></span>
<span><span class="co">#&gt;   &lt;character&gt;   &lt;character&gt;                        &lt;list&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 1    sample01 Cellbound1_z3 695 x 695 x 1 SpatRasterImage           1</span></span>
<span><span class="co">#&gt; 2    sample01 Cellbound2_z3 695 x 695 x 1 SpatRasterImage           1</span></span>
<span><span class="co">#&gt; 3    sample01 Cellbound3_z3 695 x 695 x 1 SpatRasterImage           1</span></span>
<span><span class="co">#&gt; 4    sample01       DAPI_z3 695 x 695 x 1 SpatRasterImage           1</span></span>
<span><span class="co">#&gt; 5    sample01      PolyT_z3 695 x 695 x 1 SpatRasterImage           1</span></span></code></pre></div>
<p>To see what are the <code>image_id</code>s present in the SFE
object</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/imageIDs.html" class="external-link">imageIDs</a></span><span class="op">(</span><span class="va">sfe_mer</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Cellbound1_z3" "Cellbound2_z3" "Cellbound3_z3" "DAPI_z3"      </span></span>
<span><span class="co">#&gt; [5] "PolyT_z3"</span></span></code></pre></div>
<p>Use the <code>ext</code> function to get the extent of the image</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rspatial.github.io/terra/reference/ext.html" class="external-link">ext</a></span><span class="op">(</span><span class="va">img</span><span class="op">)</span></span>
<span><span class="co">#&gt;      xmin      xmax      ymin      ymax </span></span>
<span><span class="co">#&gt;  6499.909  6800.141 -1500.166 -1199.939</span></span></code></pre></div>
<p>The images from SFE objects can be plotted with the
<code><a href="https://rdrr.io/pkg/Voyager/man/plotImage.html" class="external-link">plotImage()</a></code> function in <code>Voyager</code></p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotImage.html" class="external-link">plotImage</a></span><span class="op">(</span><span class="va">sfe_mer</span>, image_id <span class="op">=</span> <span class="st">"PolyT_z3"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-70-1.png" width="700"></p>
<p>Or for <code>SpatRasterImage</code>, with the <code>plot</code>
function in <code>terra</code> (only works in SFE 1.7.0 or later;
earlier versions need to call <code>plot(imgRaster(img))</code>)</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rspatial.github.io/terra/reference/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">img</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-71-1.png" width="700"></p>
<p>The image can be plotted behind cell level data in all
<code>Voyager</code> functions that plot geometries</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_mer</span>, features <span class="op">=</span> <span class="st">"transcript_count"</span>, </span>
<span>                   colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>,</span>
<span>                   aes_use <span class="op">=</span> <span class="st">"color"</span>, fill <span class="op">=</span> <span class="cn">NA</span>, <span class="co"># Only color by cell outline</span></span>
<span>                   image_id <span class="op">=</span> <span class="st">"PolyT_z3"</span>, dark <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-72-1.png" width="700"></p>
<p>The advantage of <code>SpatRasterImage</code> is that one can use
vector geometries such as in <code>sf</code> data frames to extract data
from the raster image. With a binary mask, the <code>terra</code>
package can convert the mask into polygons and vice versa.</p>
<p>Here we extract PolyT intensity values within each cell and see how
they relate to total transcript counts</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">polyt_values</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/extract.html" class="external-link">extract</a></span><span class="op">(</span><span class="va">img</span>, <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">cellSeg</a></span><span class="op">(</span><span class="va">sfe_mer</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">polyt_sum</span> <span class="op">&lt;-</span> <span class="va">polyt_values</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">ID</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>polyt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">mosaic_PolyT_z3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">polyt_sum</span><span class="op">$</span><span class="va">total_counts</span> <span class="op">&lt;-</span> <span class="va">sfe_mer</span><span class="op">$</span><span class="va">transcript_count</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">polyt_sum</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">polyt</span>, <span class="va">total_counts</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density_2d.html" class="external-link">geom_density2d</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"cornflowerblue"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-73-1.png" width="700"></p>
<p>This way we can relate data from the images to other cell level data,
which includes gene expression data.</p>
<p>But there are some disadvantages, such as that <code>terra</code> is
built for geography so it’s difficult to perform affine transforms of
the image (including rotation); in geography the transformation is
performed by reprojecting the map and there are standards for the
projections such as the Mercator and Robinson projections of the world
map. So when the <code>SpatRasterImage</code> is rotated, it’s converted
into <code>ExtImage</code>, which can be converted back to
<code>SpatRasterImage</code>. <code>BioFormatsImage</code> can also be
converted into <code>SpatRasterImage</code> though that goes through
<code>ExtImage</code>.</p>
<p><code>ExtImage</code> is a thin wrapper around the <a href="https://bioconductor.org/packages/release/bioc/vignettes/EBImage/inst/doc/EBImage-introduction.html#3_Image_data_representation" class="external-link"><code>Image</code></a>
class in the <code>EBImage</code> package to conform to SPE’s
requirements and to add a spatial extent. With <code>ExtImage</code>,
one can do thresholding and morphological operations. However, it’s not
merely a wrapper; it contains another essential metadata field for the
extent. When <code>BioFormatsImage</code> is loaded into memory, it
becomes <code>EBImage</code>.</p>
<p>Here we convert this image into <code>ExtImage</code>, and plot it
with the <code>EBImage</code> package; the <code><a href="https://rdrr.io/pkg/BiocGenerics/man/normalize.html" class="external-link">normalize()</a></code>
function is called because <code><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display()</a></code> expects the values to
be between 0 and 1.</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ebi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/toExtImage.html" class="external-link">toExtImage</a></span><span class="op">(</span><span class="va">img</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/normalize.html" class="external-link">normalize</a></span><span class="op">(</span><span class="va">ebi</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-74-1.png" width="700"></p>
<p>With <code>EBImage</code>, we can apply tools from the image
processing tradition, such as <a href="https://bioconductor.org/packages/release/bioc/vignettes/EBImage/inst/doc/EBImage-introduction.html#73_Morphological_operations" class="external-link">morphological
operations</a>. Here we use Otsu threasholding to automatically find a
threshold to segment the tissue followed by an opening operation to
remove small bits and closing to fill holes</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">th</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/otsu.html" class="external-link">otsu</a></span><span class="op">(</span><span class="va">ebi</span>, range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">ebi</span><span class="op">)</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">ebi</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">mask</span> <span class="op">&lt;-</span> <span class="va">ebi</span> <span class="op">&gt;</span> <span class="va">th</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="va">mask</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-75-1.png" width="700"></p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mask</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/morphology.html" class="external-link">opening</a></span><span class="op">(</span><span class="va">mask</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/morphology.html" class="external-link">closing</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="va">mask</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-76-1.png" width="700"></p>
<p>This mask can be converted into a polygon through
<code>SpatRasterImage</code> and <code><a href="https://rspatial.github.io/terra/reference/as.polygons.html" class="external-link">terra::as.polygons</a></code>. Here
the mask still has the spatial extent, which makes sure that it’s still
aligned to the geometries regardless of pixel size.</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rspatial.github.io/terra/reference/ext.html" class="external-link">ext</a></span><span class="op">(</span><span class="va">mask</span><span class="op">)</span></span>
<span><span class="co">#&gt;      xmin      xmax      ymin      ymax </span></span>
<span><span class="co">#&gt;  6499.909  6800.141 -1500.166 -1199.939</span></span></code></pre></div>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mask_spi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/toSpatRasterImage.html" class="external-link">toSpatRasterImage</a></span><span class="op">(</span><span class="va">mask</span><span class="op">)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Saving image with `.tiff` (non OME-TIFF) format:</span></span>
<span><span class="co">#&gt; img.tiff</span></span>
<span><span class="op">(</span><span class="va">tb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/as.polygons.html" class="external-link">as.polygons</a></span><span class="op">(</span><span class="va">mask_spi</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_as_sf.html" class="external-link">st_as_sf</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Simple feature collection with 2 features and 1 field</span></span>
<span><span class="co">#&gt; Geometry type: GEOMETRY</span></span>
<span><span class="co">#&gt; Dimension:     XY</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: 6499.909 ymin: -1500.166 xmax: 6800.141 ymax: -1199.939</span></span>
<span><span class="co">#&gt; CRS:           NA</span></span>
<span><span class="co">#&gt;   lyr.1                       geometry</span></span>
<span><span class="co">#&gt; 1     0 MULTIPOLYGON (((6499.909 -1...</span></span>
<span><span class="co">#&gt; 2     1 POLYGON ((6499.909 -1199.93...</span></span></code></pre></div>
<p>Here the region with value 0 (not in tissue) and value 1 (in tissue)
have been converted to polygons.</p>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">tb</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">lyr.1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-79-1.png" width="700"></p>
<p>The polygons can be added back to the SFE object; here say I’m only
interested in value 1</p>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tb</span> <span class="op">&lt;-</span> <span class="va">tb</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">lyr.1</span> <span class="op">==</span> <span class="fl">1L</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">geometry</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>sample_id <span class="op">=</span> <span class="st">"sample01"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">tissueBoundary</a></span><span class="op">(</span><span class="va">sfe_mer</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">tb</span></span></code></pre></div>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_mer</span>, features <span class="op">=</span> <span class="st">"transcript_count"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>,</span>
<span>                   annotGeometryName <span class="op">=</span> <span class="st">"tissueBoundary"</span>,</span>
<span>                   aes_use <span class="op">=</span> <span class="st">"color"</span>, fill <span class="op">=</span> <span class="cn">NA</span>, <span class="co"># Only color by cell outline</span></span>
<span>                   <span class="co"># Don't fill the tissue boundary either</span></span>
<span>                   annot_fixed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="cn">NA</span>, color <span class="op">=</span> <span class="st">"magenta"</span><span class="op">)</span>,</span>
<span>                   image_id <span class="op">=</span> <span class="st">"PolyT_z3"</span>, dark <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-81-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="bioformatsimage">
<code>BioFormatsImage</code><a class="anchor" aria-label="anchor" href="#bioformatsimage"></a>
</h3>
<p><code>BioFormatsImage</code> is used for <a href="https://docs.openmicroscopy.org/ome-model/5.6.3/ome-tiff/" class="external-link"><code>OME-TIFF</code></a>
images whose compression can’t be read by <code>terra</code>. The image
is not loaded into memory. It’s just some metadata, which includes the
file path, extent, origin (minimum value of coordinates), and affine
transformation information to apply the transformation on the fly when a
part of the image is loaded into memory. So far functions related to
<code>BioFormatsImage</code> cater to Xenium data. This is somewhat
similar to <a href="https://r-spatial.github.io/stars/articles/stars2.html" class="external-link"><code>stars</code>
proxy objects</a>, which also only has metadata in memory, in another
commonly used raster package <code>stars</code>.</p>
<p>Here images from the Xenium dataset is read as
<code>BioFormatsImage</code></p>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">imgData</a></span><span class="op">(</span><span class="va">sfe_xen</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 1 row and 4 columns</span></span>
<span><span class="co">#&gt;     sample_id         image_id                            data scaleFactor</span></span>
<span><span class="co">#&gt;   &lt;character&gt;      &lt;character&gt;                          &lt;AsIs&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 1    sample01 morphology_focus 1207 x 1186 x 4 BioFormatsImage           1</span></span></code></pre></div>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/imgData-methods.html" class="external-link">getImg</a></span><span class="op">(</span><span class="va">sfe_xen</span><span class="op">)</span></span>
<span><span class="co">#&gt; X: 1207, Y: 1186, C: 4, Z: 1, T: 1, BioFormatsImage</span></span>
<span><span class="co">#&gt; imgSource():</span></span>
<span><span class="co">#&gt;   /tmp/RtmpOVsuEd/xenium/xenium2/morphology_focus/morphology_focus_0000.ome.tif</span></span></code></pre></div>
<p>While the images in the MERFISH dataset above only have 1 channel,
this image has 4 channels, which are DAPI, ATP1A1/CD45/E-Cadherin, 18S,
and AlphaSMA/Vimentin, in this order. The images are pyramids with
multiple resolutions; some applications would not require the highest
resolution. The file for each channel is around 370 MB. Only the
metadata is read in R and only the relevant portion of the image at the
highest resolution necessary is loaded into memory when needed, say when
plotting.</p>
<p>One can select up to 3 channels (in the order of RGB) for plotting;
<code>channel</code> must be specified when there are more than 3
channels in the image, and this may not be colorblind friendly</p>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotImage.html" class="external-link">plotImage</a></span><span class="op">(</span><span class="va">sfe_xen</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>, channel <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">1</span>, </span>
<span>          normalize_channels <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-84-1.png" width="700"></p>
<p>Or plot the channels separately in gray scale</p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotImage.html" class="external-link">plotImage</a></span><span class="op">(</span><span class="va">sfe_xen</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>, channel <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotImage.html" class="external-link">plotImage</a></span><span class="op">(</span><span class="va">sfe_xen</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>, channel <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-85-1.png" width="700"></p>
<p>The color palette can be changed</p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotImage.html" class="external-link">plotImage</a></span><span class="op">(</span><span class="va">sfe_xen</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>, channel <span class="op">=</span> <span class="fl">1</span>, </span>
<span>          palette <span class="op">=</span> <span class="fu"><a href="https://scales.r-lib.org/reference/pal_viridis.html" class="external-link">viridis_pal</a></span><span class="op">(</span><span class="op">)</span><span class="op">(</span><span class="fl">255</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-86-1.png" width="700"></p>
<p>Multiple channels can be plotted behind geometries and cell level
data (the image in the example data is not the highest resolution from
the original data from the 10X website in order to reduce download
time)</p>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_xen</span>, <span class="va">genes_use</span><span class="op">[</span><span class="fl">6</span><span class="op">]</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, </span>
<span>                   rowGeometryName <span class="op">=</span> <span class="st">"txSpots"</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>,</span>
<span>                   fill <span class="op">=</span> <span class="cn">NA</span>, aes_use <span class="op">=</span> <span class="st">"color"</span>, </span>
<span>                   tx_fixed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"lightgray"</span><span class="op">)</span>,</span>
<span>                   channel <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">1</span>, bbox <span class="op">=</span> <span class="va">bbox</span>, dark <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-87-1.png" width="700"></p>
<p>Converting the image into <code>ExtImage</code> will load it into
memory, and one can select the resolution in the pyramid to load</p>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">ebi2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/toExtImage.html" class="external-link">toExtImage</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/imgData-methods.html" class="external-link">getImg</a></span><span class="op">(</span><span class="va">sfe_xen</span><span class="op">)</span>, resolution <span class="op">=</span> <span class="fl">2L</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; 1207 x 1186 x 4 (width x height x channels) ExtImage</span></span></code></pre></div>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use the arrows on the top left of the widget to see different channels</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/normalize.html" class="external-link">normalize</a></span><span class="op">(</span><span class="va">ebi2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Only the first frame of the image stack is displayed.</span></span>
<span><span class="co">#&gt; To display all frames use 'all = TRUE'.</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-89-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="spatial-operations">Spatial operations<a class="anchor" aria-label="anchor" href="#spatial-operations"></a>
</h2>
<div class="section level3">
<h3 id="bounding-box">Bounding box<a class="anchor" aria-label="anchor" href="#bounding-box"></a>
</h3>
<p>The bounding box of a geometry is the smallest rectangle that
contains this geometry, so you get minimum and maximum x coordinates and
y coordinates. We can find the bounding box of individual
<code>sf</code> data frames with <code>st_bbox</code> from the
<code>sf</code> package</p>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_bbox.html" class="external-link">st_bbox</a></span><span class="op">(</span><span class="va">rg</span><span class="op">)</span></span>
<span><span class="co">#&gt;          xmin          ymin          xmax          ymax </span></span>
<span><span class="co">#&gt;  1.159668e-03 -1.008098e+03  1.026159e+03 -4.473946e+00</span></span></code></pre></div>
<p>However, in an SFE object, there are multiple geometries, such as
cell centroids, cell segmentation, nucleus segmentation, tissue
boundary, transcript spots, and so on, and there are images. The
<code>bbox</code> function for SFE aggregates the bounding boxes of all
the geometries (and optionally images) to get an overall bounding box of
the SFE object:</p>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/bbox-SpatialFeatureExperiment-method.html" class="external-link">bbox</a></span><span class="op">(</span><span class="va">sfe_xen</span><span class="op">)</span></span>
<span><span class="co">#&gt;      xmin      ymin      xmax      ymax </span></span>
<span><span class="co">#&gt;     0.000 -1008.100  1026.163     0.000</span></span></code></pre></div>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># In this case the image is not larger than the geometries</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/bbox-SpatialFeatureExperiment-method.html" class="external-link">bbox</a></span><span class="op">(</span><span class="va">sfe_xen</span>, include_image <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt;      xmin      ymin      xmax      ymax </span></span>
<span><span class="co">#&gt;     0.000 -1008.100  1026.163     0.000</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="cropping">Cropping<a class="anchor" aria-label="anchor" href="#cropping"></a>
</h3>
<p>You can think of the SFE object as a stack of maps that are aligned,
like the <a href="https://apps.nationalmap.gov/viewer/" class="external-link">National Map</a>
layers of satellite images, land use, administrative boundaries,
watersheds, rock formations, faults, and etc. Cropping will crop all of
the maps. One can crop with either a bounding box or a polygon of any
shape. The <code>colGeometryName</code> argument specifies the
<code>colGeometry</code> to decide which cell to keep after cropping.
Using the centroid would be different from using the cell polygon since
a polygon can slightly overlap with the bounding box while the centroid
is outside.</p>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_cropped</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/crop.html" class="external-link">crop</a></span><span class="op">(</span><span class="va">sfe_xen</span>, <span class="va">bbox</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/bbox-SpatialFeatureExperiment-method.html" class="external-link">bbox</a></span><span class="op">(</span><span class="va">sfe_cropped</span><span class="op">)</span></span>
<span><span class="co">#&gt; xmin ymin xmax ymax </span></span>
<span><span class="co">#&gt;    0 -600  200 -400</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">sfe_cropped</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 394 207</span></span></code></pre></div>
<p>Now we only have 207 cells in this bounding box as opposed to around
6000 in the original. Here plot the cropped SFE object</p>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_cropped</span>, <span class="va">genes_use</span><span class="op">[</span><span class="fl">6</span><span class="op">]</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, </span>
<span>                   rowGeometryName <span class="op">=</span> <span class="st">"txSpots"</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>,</span>
<span>                   fill <span class="op">=</span> <span class="cn">NA</span>, aes_use <span class="op">=</span> <span class="st">"color"</span>, </span>
<span>                   tx_fixed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"lightgray"</span><span class="op">)</span>,</span>
<span>                   channel <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">1</span>, dark <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-94-1.png" width="700"></p>
<p>See documentation of <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/crop.html" class="external-link">SpatialFeatureExperiment::crop()</a></code>
for more options, such as only keeping cells entirely contained in the
polygon used for cropping and using a geometry to exclude rather than
include cells.</p>
</div>
<div class="section level3">
<h3 id="transformation">Transformation<a class="anchor" aria-label="anchor" href="#transformation"></a>
</h3>
<p>We can rotate (right now only multiples of 90 degrees), mirror,
transpose, and translate the SFE object, such as when there’s a
canonical orientation like in brain sections but the data is of a
different orientation when read in. Here all geometries and images are
transformed while keeping them aligned.</p>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_mirror</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/SFE-transform.html" class="external-link">mirror</a></span><span class="op">(</span><span class="va">sfe_xen</span>, direction <span class="op">=</span> <span class="st">"vertical"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_mirror</span>, <span class="va">genes_use</span><span class="op">[</span><span class="fl">6</span><span class="op">]</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, </span>
<span>                   rowGeometryName <span class="op">=</span> <span class="st">"txSpots"</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>,</span>
<span>                   fill <span class="op">=</span> <span class="cn">NA</span>, aes_use <span class="op">=</span> <span class="st">"color"</span>, linewidth <span class="op">=</span> <span class="fl">0.15</span>,</span>
<span>                   tx_fixed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"lightgray"</span><span class="op">)</span>,</span>
<span>                   channel <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">1</span>, dark <span class="op">=</span> <span class="cn">TRUE</span>, normalize_channels <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-96-1.png" width="700"></p>
<p>“Mirror” is one of the named transformations; other transformations
include scale, transpose, rotate, and translate.</p>
<p>More generally, we can use a transformation matrix for affine
transformation, say to align this dataset to another dataset of a
different modality from a serial section. We can obtain this
transformation matrix from <a href="https://imagej.net/plugins/bigwarp" class="external-link">BigWarp</a> in ImageJ for
example, where one can manually annotate corresponding points in the
adjacent sections. Here matrix <code>M</code> performs the linear
transformation, and <code>v</code> is translation. The transformation
can only be performed in the x-y plane at present. Contributions to
allow non-linear transformation are welcome.</p>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Here it's a rotation matrix but can be anything</span></span>
<span><span class="va">M</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Trig.html" class="external-link">cos</a></span><span class="op">(</span><span class="va">pi</span><span class="op">/</span><span class="fl">6</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Trig.html" class="external-link">sin</a></span><span class="op">(</span><span class="va">pi</span><span class="op">/</span><span class="fl">6</span><span class="op">)</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Trig.html" class="external-link">sin</a></span><span class="op">(</span><span class="va">pi</span><span class="op">/</span><span class="fl">6</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Trig.html" class="external-link">cos</a></span><span class="op">(</span><span class="va">pi</span><span class="op">/</span><span class="fl">6</span><span class="op">)</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">sfe_affine</span> <span class="op">&lt;-</span> <span class="fu">SpatialFeatureExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/SFE-transform.html" class="external-link">affine</a></span><span class="op">(</span><span class="va">sfe_xen</span>, M <span class="op">=</span> <span class="va">M</span>, v <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_affine</span>, <span class="va">genes_use</span><span class="op">[</span><span class="fl">6</span><span class="op">]</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, </span>
<span>                   rowGeometryName <span class="op">=</span> <span class="st">"txSpots"</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>,</span>
<span>                   fill <span class="op">=</span> <span class="cn">NA</span>, aes_use <span class="op">=</span> <span class="st">"color"</span>, linewidth <span class="op">=</span> <span class="fl">0.15</span>,</span>
<span>                   tx_fixed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"lightgray"</span><span class="op">)</span>,</span>
<span>                   channel <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">1</span>, dark <span class="op">=</span> <span class="cn">TRUE</span>, normalize_channels <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-98-1.png" width="700"></p>
<p>What happens to the image after transformation? It’s still a
BioFormatsImage, which means there’s only metadata in memory.</p>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/imgData-methods.html" class="external-link">getImg</a></span><span class="op">(</span><span class="va">sfe_affine</span><span class="op">)</span></span>
<span><span class="co">#&gt; X: 1207, Y: 1186, C: 4, Z: 1, T: 1, BioFormatsImage</span></span>
<span><span class="co">#&gt; imgSource():</span></span>
<span><span class="co">#&gt;   /tmp/RtmpOVsuEd/xenium/xenium2/morphology_focus/morphology_focus_0000.ome.tif</span></span></code></pre></div>
<p>The affine transformation information is stored in the metadata and
can be queried with <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/BioFormatsImage-getters.html" class="external-link">transformation()</a></code></p>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/BioFormatsImage-getters.html" class="external-link">transformation</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/imgData-methods.html" class="external-link">getImg</a></span><span class="op">(</span><span class="va">sfe_affine</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; $M</span></span>
<span><span class="co">#&gt;           [,1]       [,2]</span></span>
<span><span class="co">#&gt; [1,] 0.8660254 -0.5000000</span></span>
<span><span class="co">#&gt; [2,] 0.5000000  0.8660254</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $v</span></span>
<span><span class="co">#&gt; [1] 0 0</span></span></code></pre></div>
<p>This is similar to <a href="https://r-spatial.github.io/stars/reference/st_geotransform.html" class="external-link"><code>st_geotransform</code>
in the <code>stars</code> package</a> which puts the transformation in
the metadata, not resampling the image. Here in SFE, the image is only
resampled when loaded into memory as <code>ExtImage</code>.</p>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotImage.html" class="external-link">plotImage</a></span><span class="op">(</span><span class="va">sfe_affine</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>, channel <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">1</span>, </span>
<span>          normalize_channels <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-101-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="spatial-aggregation">Spatial aggregation<a class="anchor" aria-label="anchor" href="#spatial-aggregation"></a>
</h3>
<p>Imaging based technologies have single cell resolution, while the
widely used Visium does not. When a spatial neighborhood graph is used
for ESDA, the neighbors of single cells mean something different from
neighbors of Visium spots. In addition, as shown in the <a href="https://pachterlab.github.io/voyager/articles/vig6_merfish.html#morans-i" class="external-link">Voyager
MERFISH vignette</a>, a negative Moran’s I at single cell level can
become positive after spatial binning. Negative spatial autocorrelation
can coexist locally with positive spatial autocorrelation at a longer
length scale. The <a href="https://pachterlab.github.io/voyager/articles/correlogram_landing.html" class="external-link">correlogram
based on higher order neighbors</a> is one way to explore how spatial
correlation changes through length scales, though it (at least the
<code>spdep</code> implementation) is pretty slow to run and doesn’t
scale well to hundreds of thousands of cells.</p>
<p>In the <a href="https://academic.oup.com/bioinformatics/article/40/7/btae412/7696710" class="external-link"><code>SEraster</code>
package</a>, spatial aggregation is used to greatly speed up finding
spatially variable genes with <a href="https://www.nature.com/articles/s41467-023-39748-z" class="external-link"><code>nnSVG</code></a>
for large datasets while largely retaining performance at the single
cell resolution.</p>
<p>SFE (1.7.0 and later) implements spatial aggregation, not only with a
grid (square and hexagonal) defined by grid bin size but also with any
kind of geometry. Aggregation can be done at either the transcript level
(number of transcript spots per bin) or the cell level (use a summary
function such as sum or mean to summarize cell level data for cells that
intersect or are covered by the bin). This implementation is more
general than that in <code>SEraster</code>.</p>
<div class="section level4">
<h4 id="transcript-level">Transcript level<a class="anchor" aria-label="anchor" href="#transcript-level"></a>
</h4>
<p>One can create an SFE object directly from the transcript spot file
from the output of the commercial technology (not yet converted to
GeoParquet); here use hexagonal bins, size 20 microns:</p>
<div class="sourceCode" id="cb101"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/XeniumOutput.html" class="external-link">XeniumOutput</a></span><span class="op">(</span><span class="st">"v2"</span>, file_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fp</span>, <span class="st">"xenium"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; The downloaded files are in /tmp/RtmpOVsuEd/xenium/xenium2</span></span>
<span><span class="va">sfe_agg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/aggregateTx.html" class="external-link">aggregateTxTech</a></span><span class="op">(</span><span class="va">data_dir</span>, tech <span class="op">=</span> <span class="st">"Xenium"</span>, cellsize <span class="op">=</span> <span class="fl">20</span>, </span>
<span>                           square <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">sfe_agg</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 398 3068 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(398): ABCC11 ACE2 ... VWA5A VWF</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(3068): 30 31 ... 3096 3097</span></span>
<span><span class="co">#&gt; colData names(1): sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : X Y</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: micron</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: bins (POLYGON) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
<p>By default, the new <code>colGeometry</code> of the spatial bins is
called “bins”, but that can be changed with the
<code>new_geometry_name</code> argument.</p>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_agg</span><span class="op">$</span><span class="va">total_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe_agg</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_agg</span>, <span class="st">"total_counts"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-103-1.png" width="700"></p>
<p>We can also aggregate from <code>rowGeometry</code> of an SFE object,
here with square 50 micron bins</p>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_agg2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span><span class="va">sfe_xen</span>, rowGeometryName <span class="op">=</span> <span class="st">"txSpots"</span>, cellsize <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_agg2</span><span class="op">$</span><span class="va">total_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe_agg2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_agg2</span>, <span class="st">"total_counts"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-105-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="esda-at-different-length-scales">ESDA at different length scales<a class="anchor" aria-label="anchor" href="#esda-at-different-length-scales"></a>
</h4>
<p>In order not to repeated load the same transcript file into memory
when doing the aggregation for multiple bin sizes, we can load it into
memory first as a data frame.</p>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tx_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://arrow.apache.org/docs/r/reference/read_parquet.html" class="external-link">read_parquet</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"transcripts.parquet"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">areas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">500</span>,<span class="fl">10000</span>, by <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">binsizes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">areas</span><span class="op">)</span></span>
<span><span class="va">sfes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">binsizes</span>, <span class="kw">function</span><span class="op">(</span><span class="va">s</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/aggregateTx.html" class="external-link">aggregateTxTech</a></span><span class="op">(</span><span class="va">data_dir</span>, df <span class="op">=</span> <span class="va">tx_df</span>, tech <span class="op">=</span> <span class="st">"Xenium"</span>,</span>
<span>                                                     cellsize <span class="op">=</span> <span class="va">s</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Moran’s I is the most commonly used metric for spatial
autocorrelation. We can compute Moran’s I on one of the features, say
total counts; the same can be done for gene expression. Here we compute
Moran’s I using different spatial neighborhood graphs, from rook and
from queen. Rook means two polygons are only adjacent if they touch by
an edge, while queen means they can be adjacent if they touch by only a
point, so queen includes diagonal neighbors of the square grid while
rook does not.</p>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">sfes</span>, <span class="kw">function</span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">sfe</span><span class="op">$</span><span class="va">total_count</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"rook"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findSpatialNeighbors-SpatialFeatureExperiment-method.html" class="external-link">findSpatialNeighbors</a></span><span class="op">(</span><span class="va">sfe</span>, type <span class="op">=</span> <span class="st">"bins"</span>, method <span class="op">=</span> <span class="st">"poly2nb"</span>,</span>
<span>                                                  queen <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"queen"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findSpatialNeighbors-SpatialFeatureExperiment-method.html" class="external-link">findSpatialNeighbors</a></span><span class="op">(</span><span class="va">sfe</span>, type <span class="op">=</span> <span class="st">"bins"</span>, method <span class="op">=</span> <span class="st">"poly2nb"</span>,</span>
<span>                                                   queen <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>    <span class="va">sfe</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>What the smallest bins look like</p>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfes</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="st">"total_count"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-108-1.png" width="700"></p>
<p>What the largest bins look like</p>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfes</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">sfes</span><span class="op">)</span><span class="op">]</span><span class="op">]</span>, <span class="st">"total_count"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-109-1.png" width="700"></p>
<p>The results for <code>colData</code> variables are in the
<code>colFeatureData</code> field.</p>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">sfes</span>, <span class="va">colDataMoransI</span>, colGraphName <span class="op">=</span> <span class="st">"rook"</span>, feature <span class="op">=</span> <span class="st">"total_count"</span><span class="op">)</span></span>
<span><span class="va">sfes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">sfes</span>, <span class="va">colDataMoransI</span>, colGraphName <span class="op">=</span> <span class="st">"queen"</span>, feature <span class="op">=</span> <span class="st">"total_count"</span>,</span>
<span>               name <span class="op">=</span> <span class="st">"moran_queen"</span><span class="op">)</span> </span>
<span><span class="co"># Use name argument to store results under non-default name</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colFeatureData.html" class="external-link">colFeatureData</a></span><span class="op">(</span><span class="va">sfes</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 2 rows and 3 columns</span></span>
<span><span class="co">#&gt;             moran_sample01 K_sample01 moran_queen_sample01</span></span>
<span><span class="co">#&gt;                  &lt;numeric&gt;  &lt;numeric&gt;            &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; sample_id               NA         NA                   NA</span></span>
<span><span class="co">#&gt; total_count       0.371837    5.64398             0.275386</span></span></code></pre></div>
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">morans_rook</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">sfes</span>, </span>
<span>                      <span class="kw">function</span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span> </span>
<span>                          <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colFeatureData.html" class="external-link">colFeatureData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="st">"total_count"</span>, <span class="st">"moran_sample01"</span><span class="op">]</span>, </span>
<span>                      FUN.VALUE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">numeric</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">morans_queen</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">sfes</span>, </span>
<span>                      <span class="kw">function</span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span> </span>
<span>                          <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colFeatureData.html" class="external-link">colFeatureData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="st">"total_count"</span>, <span class="st">"moran_queen_sample01"</span><span class="op">]</span>, </span>
<span>                      FUN.VALUE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">numeric</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># colorblind friendly palette</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ditto_colors"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df_morans</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>rook <span class="op">=</span> <span class="va">morans_rook</span>,</span>
<span>                        queen <span class="op">=</span> <span class="va">morans_queen</span>,</span>
<span>                        sizes <span class="op">=</span> <span class="va">binsizes</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">sizes</span>, names_to <span class="op">=</span> <span class="st">"type"</span>, values_to <span class="op">=</span> <span class="st">"moran"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df_morans</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">sizes</span>, <span class="va">moran</span>, color <span class="op">=</span> <span class="va">type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">ditto_colors</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-113-1.png" width="700"></p>
<p>Here rook tends to have higher Moran’s I than queen, which makes
sense when the center of the diagonal neighbors are further away.
Spatial autocorrelation in total counts is positive at all scales. It
decreases with longer scale at first, but then increases again.</p>
<p>It would be interesting to compare bivariate such as <a href="https://pachterlab.github.io/voyager/articles/lee_landing.html" class="external-link">Lee’s
L</a> and multivariate such as <a href="https://pachterlab.github.io/voyager/articles/multispati_landing.html" class="external-link">MULTISPATI
PCA</a> spatial statistics across scales. It would also be interesting
to compare results with hexagonal grids and rotations of the grid for
more robust comparisons like in the <code>SEraster</code> paper, and to
systematically investigate the implications of different spatial
neighborhood graphs and different edge weights on downstream spatial
analyses.</p>
</div>
<div class="section level4">
<h4 id="cell-level">Cell level<a class="anchor" aria-label="anchor" href="#cell-level"></a>
</h4>
<p>We can also spatially aggregate data at the cell level, say when
transcript spots are unavailable or something other than transcript
count is desired for aggregation. Here we can aggregate by summing the
transcript counts of cells whose centroids fall into a bin, or use the
mean instead. Other functions are allowed, as long as they take a
numeric matrix as input and returns a numeric vector the same length as
the number of rows of the input matrix, such as
<code>rowMedians</code>.</p>
<div class="sourceCode" id="cb113"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_sum</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span><span class="va">sfe_xen</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, cellsize <span class="op">=</span> <span class="fl">40</span>,</span>
<span>                     square <span class="op">=</span> <span class="cn">FALSE</span>, FUN <span class="op">=</span> <span class="va">sum</span><span class="op">)</span></span>
<span><span class="va">sfe_mean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span><span class="va">sfe_xen</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, cellsize <span class="op">=</span> <span class="fl">40</span>,</span>
<span>                     square <span class="op">=</span> <span class="cn">FALSE</span>, FUN <span class="op">=</span> <span class="va">mean</span><span class="op">)</span></span>
<span><span class="va">sfe_median</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span><span class="va">sfe_xen</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, cellsize <span class="op">=</span> <span class="fl">40</span>,</span>
<span>                     square <span class="op">=</span> <span class="cn">FALSE</span>, FUN <span class="op">=</span> <span class="va">rowMedians</span><span class="op">)</span></span></code></pre></div>
<p>Note that the code is a lot faster to run when using sum or mean
because behind the scene, matrix multiplication is used to compute the
sum or mean at each bin while for other functions we need to loop
through each bin, but that can be parallelized with the
<code>BPPARAM</code> argument. Let’s plot one of the genes in space to
see the implication of choosing different aggregation functions</p>
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotCellBin2D.html" class="external-link">plotCellBin2D</a></span><span class="op">(</span><span class="va">sfe_xen</span>, binwidth <span class="op">=</span> <span class="fl">40</span>, hex <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Cell density"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_sum</span>, <span class="va">genes_use</span><span class="op">[</span><span class="fl">6</span><span class="op">]</span>, exprs_values <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Sum"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_mean</span>, <span class="va">genes_use</span><span class="op">[</span><span class="fl">6</span><span class="op">]</span>, exprs_values <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Mean"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_median</span>, <span class="va">genes_use</span><span class="op">[</span><span class="fl">6</span><span class="op">]</span>, exprs_values <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Median"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span> <span class="op">+</span> <span class="va">p3</span> <span class="op">+</span> <span class="va">p4</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-115-1.png" width="700"></p>
<p>Note that the sum aggregation looks a bit more like cell density than
mean and median aggregation, because the cells are not evenly
distributed in space and when there are more cells in a bin, there are
more likely to be more transcript counts as well. Also, when using the
sum, the bins at the edge of the tissue that are not full are more
likely to have lower transcript counts.</p>
<p>Aggregation by any geometry, not just spatial grids, may facilitate
comparison of data from adjacent sections and different modalities, say
when we have Xenium spatial transcriptomics at single molecule
resolution and MALDI-MSI spatial metabolomics data with 20 micron
resolution, or when we have both Visium and GeoMX DSP. There’re some
questions to think more carefully about. The choice of transcript and
cell level and function used to aggregate cell level data would affect
the appropriate data normalization. The choice of function used to
aggregate cell level data would also have implications to downstream
analyses. Also, what to do when we only have cell level data and a cell
partially overlaps with a bin? The problem of having data at overlapping
extent but with different resolutions over areas that may or may not
overlap is called “spatial misalignment”, and there is a tradition of <a href="https://www.taylorfrancis.com/chapters/mono/10.1201/b17115-11/spatial-misalignment-sudipto-banerjee-bradley-carlin-alan-gelfand?context=ubx&amp;refId=bf6510b1-b6b6-4a84-b919-c54bd609663a" class="external-link">Bayesian
modeling for such data</a> from geography.</p>
</div>
</div>
<div class="section level3">
<h3 id="split">Split<a class="anchor" aria-label="anchor" href="#split"></a>
</h3>
<p>The opposite of aggregate would be split. While it’s possible to
split SFE objects with the grid like in the previous section, which can
be interesting as it would be interesting to compute global spatial
statistics on the separate SFE object at each bin as some sort of
local-ish spatial statistics, a more common application may be to
separate different histological regions or different pieces of tissues
imaged together, such as in a tissue microarray (TMA) which has been
applied to <a href="https://doi.org/10.1101/2024.05.31.596759" class="external-link">Visium</a>, <a href="https://doi.org/10.1101/2023.12.07.570603" class="external-link">MERFISH, Xenium, and
CosMX</a>.</p>
<p>We can split by polygons that indicate different pieces of tissue;
cells that intersect with different polygons will go to different SFE
objects</p>
<div class="sourceCode" id="cb115"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">pieces</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-116-1.png" width="700"></p>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfes_pieces</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/splitByCol.html" class="external-link">splitByCol</a></span><span class="op">(</span><span class="va">sfe_xen</span>, <span class="va">pieces</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb117"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Combine them back as different samples</span></span>
<span><span class="va">sfes_pieces</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/changeSampleIDs.html" class="external-link">changeSampleIDs</a></span><span class="op">(</span><span class="va">sfes_pieces</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>sample01 <span class="op">=</span> <span class="st">"sample02"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sfe2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">cbind</span>, <span class="va">sfes_pieces</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotGeometry.html" class="external-link">plotGeometry</a></span><span class="op">(</span><span class="va">sfe2</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-118-1.png" width="700"></p>
<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotColGraph.html" class="external-link">plotColGraph</a></span><span class="op">(</span><span class="va">sfe2</span>, <span class="st">"knn"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-119-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="esda-to-compare-samples">ESDA to compare samples<a class="anchor" aria-label="anchor" href="#esda-to-compare-samples"></a>
</h3>
<p>The two samples here are really from the same piece of tissue because
the original tissue is 3 dimensional and the blood vessel is but a
tunnel in the 3D tissue. But what if some genes have different levels of
spatial autocorrelation across the two pieces? This is a slightly more
local Moran’s I than computing Moran’s I for the whole piece, and
strength of spatial autocorrelation can have a lot of heterogeneity. It
would be more interesting if the two samples are from different
biological conditions like healthy vs. pathological. Here this is just a
quick and dirty comparison and not statistically rigorous. It merely
indicates something interesting, to be further confirmed.</p>
<p>Because the data is in <code>DelayedArray</code> and not fully loaded
into memory, computing Moran’s I for all genes would be slower than for
<code>dgCMatrix</code> in memory.</p>
<div class="sourceCode" id="cb119"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfes_pieces</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">sfes_pieces</span>, <span class="va">runMoransI</span>, colGraphName <span class="op">=</span> <span class="st">"knn"</span><span class="op">)</span></span></code></pre></div>
<p>For gene expression, global spatial results are stored in
<code>rowData</code>.</p>
<div class="sourceCode" id="cb120"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfes_pieces</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 394 rows and 5 columns</span></span>
<span><span class="co">#&gt;                                             ID                 Symbol</span></span>
<span><span class="co">#&gt;                                    &lt;character&gt;            &lt;character&gt;</span></span>
<span><span class="co">#&gt; ABCC11                         ENSG00000121270                 ABCC11</span></span>
<span><span class="co">#&gt; ACE2                           ENSG00000130234                   ACE2</span></span>
<span><span class="co">#&gt; ACKR1                          ENSG00000213088                  ACKR1</span></span>
<span><span class="co">#&gt; ACTA2                          ENSG00000107796                  ACTA2</span></span>
<span><span class="co">#&gt; ACTG2                          ENSG00000163017                  ACTG2</span></span>
<span><span class="co">#&gt; ...                                        ...                    ...</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0461 UnassignedCodeword_0.. UnassignedCodeword_0..</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0469 UnassignedCodeword_0.. UnassignedCodeword_0..</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0479 UnassignedCodeword_0.. UnassignedCodeword_0..</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0488 UnassignedCodeword_0.. UnassignedCodeword_0..</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0497 UnassignedCodeword_0.. UnassignedCodeword_0..</span></span>
<span><span class="co">#&gt;                                        Type moran_sample01 K_sample01</span></span>
<span><span class="co">#&gt;                                 &lt;character&gt;      &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ABCC11                      Gene Expression    -0.00148686  791.96123</span></span>
<span><span class="co">#&gt; ACE2                        Gene Expression     0.34077210   23.85287</span></span>
<span><span class="co">#&gt; ACKR1                       Gene Expression     0.13361491  100.81951</span></span>
<span><span class="co">#&gt; ACTA2                       Gene Expression     0.11683337    9.60491</span></span>
<span><span class="co">#&gt; ACTG2                       Gene Expression     0.18008498    8.02017</span></span>
<span><span class="co">#&gt; ...                                     ...            ...        ...</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0461 Unassigned Codeword            NaN        NaN</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0469 Unassigned Codeword            NaN        NaN</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0479 Unassigned Codeword   -0.000474158       2108</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0488 Unassigned Codeword            NaN        NaN</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0497 Unassigned Codeword            NaN        NaN</span></span></code></pre></div>
<p>Global spatial statistics return one result for the entire sample,
regardless of local heterogeneity. Local spatial statistics such as <a href="https://pachterlab.github.io/voyager/articles/localmoran_landing.html" class="external-link">local
Moran’s I</a> return one result for each location such as cell or bin or
spot. The results are stored in the <code>localResults</code> field in
SFE objects, whose getters and setters are similar to those of
<code>reducedDims</code> and whose plotting function
<code><a href="https://rdrr.io/pkg/Voyager/man/plotLocalResult.html" class="external-link">plotLocalResult()</a></code> is very similar to
<code><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature()</a></code>.</p>
<div class="sourceCode" id="cb121"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">morans_pieces</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfes_pieces</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Symbol"</span>, <span class="st">"Type"</span>, <span class="st">"moran_sample01"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                       <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfes_pieces</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span>,<span class="st">"moran_sample02"</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rspatial.github.io/terra/reference/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">morans_pieces</span> <span class="op">&lt;-</span> <span class="va">morans_pieces</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Type</span> <span class="op">==</span> <span class="st">"Gene Expression"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb122"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">morans_pieces</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">moran_sample01</span>, <span class="va">moran_sample02</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>slope <span class="op">=</span> <span class="fl">1</span>, intercept <span class="op">=</span> <span class="fl">0</span>, color <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density_2d.html" class="external-link">geom_density2d</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"cornflowerblue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_equal</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-123-1.png" width="700"></p>
<p>The <a href="https://www.alexejgossmann.com/ccc/" class="external-link">concordance
correlation coefficient (CCC)</a> shows how well the bivariate
distribution follows the y = x line.</p>
<div class="sourceCode" id="cb123"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">metrica</span><span class="fu">::</span><span class="fu"><a href="https://adriancorrendo.github.io/metrica/reference/CCC.html" class="external-link">CCC</a></span><span class="op">(</span>obs <span class="op">=</span> <span class="va">morans_pieces</span><span class="op">$</span><span class="va">moran_sample01</span>, pred <span class="op">=</span> <span class="va">morans_pieces</span><span class="op">$</span><span class="va">moran_sample02</span><span class="op">)</span></span>
<span><span class="co">#&gt; $CCC</span></span>
<span><span class="co">#&gt; [1] 0.8873082</span></span></code></pre></div>
<p>As expected, since the two pieces are in fact one, the genes mostly
have similar Moran’s I in both pieces. It would be cool to test whether
difference in Moran’s I in the two pieces is significant, just like
testing whether the mean is different in a t-test. Whether this
difference is significant or even meaningful would also depend on the
heterogeneity in spatial structure in the same tissue type in different
2D sections of the original 3D tissue, the size of the pieces, and
questions asked; these are relevant to the variance of Moran’s I in
pieces of this size in this tissue type and variance is relevant just
like in t-tests.</p>
<p><img src="workshop_files/figure-html/unnamed-chunk-126-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb124"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8          LC_NUMERIC=C                 </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8           LC_COLLATE=en_US.UTF-8       </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8       LC_MESSAGES=en_US.UTF-8      </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8          LC_NAME=en_US.UTF-8          </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=en_US.UTF-8        LC_TELEPHONE=en_US.UTF-8     </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8    LC_IDENTIFICATION=en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Etc/UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] metrica_2.1.0                  sparseMatrixStats_1.17.2      </span></span>
<span><span class="co">#&gt;  [3] arrow_16.1.0                   stringr_1.5.1                 </span></span>
<span><span class="co">#&gt;  [5] tibble_3.2.1                   tidyr_1.3.1                   </span></span>
<span><span class="co">#&gt;  [7] dplyr_1.1.4                    patchwork_1.2.0               </span></span>
<span><span class="co">#&gt;  [9] scales_1.3.0                   EBImage_4.47.0                </span></span>
<span><span class="co">#&gt; [11] Voyager_1.7.0                  SFEData_1.7.0                 </span></span>
<span><span class="co">#&gt; [13] SpatialFeatureExperiment_1.7.1 SpatialExperiment_1.15.1      </span></span>
<span><span class="co">#&gt; [15] scater_1.33.4                  scuttle_1.15.2                </span></span>
<span><span class="co">#&gt; [17] Seurat_5.1.0                   SeuratObject_5.0.2            </span></span>
<span><span class="co">#&gt; [19] sp_2.1-4                       SingleCellExperiment_1.27.2   </span></span>
<span><span class="co">#&gt; [21] SummarizedExperiment_1.35.1    Biobase_2.65.0                </span></span>
<span><span class="co">#&gt; [23] GenomicRanges_1.57.1           GenomeInfoDb_1.41.1           </span></span>
<span><span class="co">#&gt; [25] IRanges_2.39.2                 S4Vectors_0.43.2              </span></span>
<span><span class="co">#&gt; [27] BiocGenerics_0.51.0            MatrixGenerics_1.17.0         </span></span>
<span><span class="co">#&gt; [29] matrixStats_1.3.0              ggplot2_3.5.1                 </span></span>
<span><span class="co">#&gt; [31] terra_1.7-78                   fs_1.6.4                      </span></span>
<span><span class="co">#&gt; [33] sf_1.0-16                      Matrix_1.7-0                  </span></span>
<span><span class="co">#&gt; [35] BiocStyle_2.33.1              </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] R.methodsS3_1.8.2         tiff_0.1-12              </span></span>
<span><span class="co">#&gt;   [3] RBioFormats_1.5.1         goftest_1.2-3            </span></span>
<span><span class="co">#&gt;   [5] Biostrings_2.73.1         HDF5Array_1.33.5         </span></span>
<span><span class="co">#&gt;   [7] vctrs_0.6.5               spatstat.random_3.3-1    </span></span>
<span><span class="co">#&gt;   [9] energy_1.7-11             digest_0.6.36            </span></span>
<span><span class="co">#&gt;  [11] png_0.1-8                 proxy_0.4-27             </span></span>
<span><span class="co">#&gt;  [13] ggrepel_0.9.5             deldir_2.0-4             </span></span>
<span><span class="co">#&gt;  [15] parallelly_1.37.1         magick_2.8.4             </span></span>
<span><span class="co">#&gt;  [17] MASS_7.3-61               pkgdown_2.1.0            </span></span>
<span><span class="co">#&gt;  [19] reshape2_1.4.4            httpuv_1.6.15            </span></span>
<span><span class="co">#&gt;  [21] scico_1.5.0               withr_3.0.0              </span></span>
<span><span class="co">#&gt;  [23] xfun_0.46                 survival_3.7-0           </span></span>
<span><span class="co">#&gt;  [25] memoise_2.0.1             s2_1.1.7                 </span></span>
<span><span class="co">#&gt;  [27] hexbin_1.28.3             ggbeeswarm_0.7.2         </span></span>
<span><span class="co">#&gt;  [29] systemfonts_1.1.0         ragg_1.3.2               </span></span>
<span><span class="co">#&gt;  [31] zoo_1.8-12                pbapply_1.7-2            </span></span>
<span><span class="co">#&gt;  [33] R.oo_1.26.0               spData_2.3.1             </span></span>
<span><span class="co">#&gt;  [35] KEGGREST_1.45.1           promises_1.3.0           </span></span>
<span><span class="co">#&gt;  [37] httr_1.4.7                globals_0.16.3           </span></span>
<span><span class="co">#&gt;  [39] fitdistrplus_1.2-1        rhdf5filters_1.17.0      </span></span>
<span><span class="co">#&gt;  [41] rhdf5_2.49.0              UCSC.utils_1.1.0         </span></span>
<span><span class="co">#&gt;  [43] units_0.8-5               miniUI_0.1.1.1           </span></span>
<span><span class="co">#&gt;  [45] generics_0.1.3            curl_5.2.1               </span></span>
<span><span class="co">#&gt;  [47] zlibbioc_1.51.1           ScaledMatrix_1.13.0      </span></span>
<span><span class="co">#&gt;  [49] polyclip_1.10-6           GenomeInfoDbData_1.2.12  </span></span>
<span><span class="co">#&gt;  [51] ExperimentHub_2.13.0      SparseArray_1.5.25       </span></span>
<span><span class="co">#&gt;  [53] fftwtools_0.9-11          xtable_1.8-4             </span></span>
<span><span class="co">#&gt;  [55] desc_1.4.3                evaluate_0.24.0          </span></span>
<span><span class="co">#&gt;  [57] S4Arrays_1.5.5            BiocFileCache_2.13.0     </span></span>
<span><span class="co">#&gt;  [59] bookdown_0.40             irlba_2.3.5.1            </span></span>
<span><span class="co">#&gt;  [61] sfarrow_0.4.1             colorspace_2.1-0         </span></span>
<span><span class="co">#&gt;  [63] filelock_1.0.3            isoband_0.2.7            </span></span>
<span><span class="co">#&gt;  [65] ROCR_1.0-11               reticulate_1.38.0        </span></span>
<span><span class="co">#&gt;  [67] spatstat.data_3.1-2       magrittr_2.0.3           </span></span>
<span><span class="co">#&gt;  [69] lmtest_0.9-40             spdep_1.3-5              </span></span>
<span><span class="co">#&gt;  [71] later_1.3.2               viridis_0.6.5            </span></span>
<span><span class="co">#&gt;  [73] lattice_0.22-6            spatstat.geom_3.3-2      </span></span>
<span><span class="co">#&gt;  [75] future.apply_1.11.2       scattermore_1.2          </span></span>
<span><span class="co">#&gt;  [77] cowplot_1.1.3             RcppAnnoy_0.0.22         </span></span>
<span><span class="co">#&gt;  [79] class_7.3-22              pillar_1.9.0             </span></span>
<span><span class="co">#&gt;  [81] nlme_3.1-165              compiler_4.4.1           </span></span>
<span><span class="co">#&gt;  [83] beachmat_2.21.4           RSpectra_0.16-2          </span></span>
<span><span class="co">#&gt;  [85] stringi_1.8.4             tensor_1.5               </span></span>
<span><span class="co">#&gt;  [87] plyr_1.8.9                crayon_1.5.3             </span></span>
<span><span class="co">#&gt;  [89] abind_1.4-5               locfit_1.5-9.10          </span></span>
<span><span class="co">#&gt;  [91] bit_4.0.5                 codetools_0.2-20         </span></span>
<span><span class="co">#&gt;  [93] textshaping_0.4.0         BiocSingular_1.21.2      </span></span>
<span><span class="co">#&gt;  [95] bslib_0.7.0               e1071_1.7-14             </span></span>
<span><span class="co">#&gt;  [97] plotly_4.10.4             mime_0.12                </span></span>
<span><span class="co">#&gt;  [99] splines_4.4.1             Rcpp_1.0.13              </span></span>
<span><span class="co">#&gt; [101] fastDummies_1.7.3         dbplyr_2.5.0             </span></span>
<span><span class="co">#&gt; [103] knitr_1.48                blob_1.2.4               </span></span>
<span><span class="co">#&gt; [105] utf8_1.2.4                BiocVersion_3.20.0       </span></span>
<span><span class="co">#&gt; [107] listenv_0.9.1             DelayedMatrixStats_1.27.2</span></span>
<span><span class="co">#&gt; [109] gsl_2.1-8                 statmod_1.5.0            </span></span>
<span><span class="co">#&gt; [111] pkgconfig_2.0.3           tools_4.4.1              </span></span>
<span><span class="co">#&gt; [113] cachem_1.1.0              RSQLite_2.3.7            </span></span>
<span><span class="co">#&gt; [115] viridisLite_0.4.2         DBI_1.2.3                </span></span>
<span><span class="co">#&gt; [117] minerva_1.5.10            fastmap_1.2.0            </span></span>
<span><span class="co">#&gt; [119] rmarkdown_2.27            grid_4.4.1               </span></span>
<span><span class="co">#&gt; [121] ica_1.0-3                 AnnotationHub_3.13.0     </span></span>
<span><span class="co">#&gt; [123] sass_0.4.9                BiocManager_1.30.23      </span></span>
<span><span class="co">#&gt; [125] dotCall64_1.1-1           RANN_2.6.1               </span></span>
<span><span class="co">#&gt; [127] farver_2.1.2              wk_0.9.2                 </span></span>
<span><span class="co">#&gt; [129] yaml_2.3.9                cli_3.6.3                </span></span>
<span><span class="co">#&gt; [131] purrr_1.0.2               leiden_0.4.3.1           </span></span>
<span><span class="co">#&gt; [133] lifecycle_1.0.4           uwot_0.2.2               </span></span>
<span><span class="co">#&gt; [135] bluster_1.15.0            DropletUtils_1.25.0      </span></span>
<span><span class="co">#&gt; [137] BiocParallel_1.39.0       gtable_0.3.5             </span></span>
<span><span class="co">#&gt; [139] rjson_0.2.21              ggridges_0.5.6           </span></span>
<span><span class="co">#&gt; [141] progressr_0.14.0          parallel_4.4.1           </span></span>
<span><span class="co">#&gt; [143] limma_3.61.5              jsonlite_1.8.8           </span></span>
<span><span class="co">#&gt; [145] edgeR_4.3.5               RcppHNSW_0.6.0           </span></span>
<span><span class="co">#&gt; [147] bitops_1.0-7              bit64_4.0.5              </span></span>
<span><span class="co">#&gt; [149] assertthat_0.2.1          Rtsne_0.17               </span></span>
<span><span class="co">#&gt; [151] spatstat.utils_3.0-5      BiocNeighbors_1.23.0     </span></span>
<span><span class="co">#&gt; [153] jquerylib_0.1.4           highr_0.11               </span></span>
<span><span class="co">#&gt; [155] dqrng_0.4.1               zeallot_0.1.0            </span></span>
<span><span class="co">#&gt; [157] spatstat.univar_3.0-0     R.utils_2.12.3           </span></span>
<span><span class="co">#&gt; [159] lazyeval_0.2.2            shiny_1.8.1.1            </span></span>
<span><span class="co">#&gt; [161] htmltools_0.5.8.1         rJava_1.0-11             </span></span>
<span><span class="co">#&gt; [163] sctransform_0.4.1         rappdirs_0.3.3           </span></span>
<span><span class="co">#&gt; [165] glue_1.7.0                spam_2.10-0              </span></span>
<span><span class="co">#&gt; [167] XVector_0.45.0            RCurl_1.98-1.16          </span></span>
<span><span class="co">#&gt; [169] classInt_0.4-10           jpeg_0.1-10              </span></span>
<span><span class="co">#&gt; [171] gridExtra_2.3             boot_1.3-30              </span></span>
<span><span class="co">#&gt; [173] igraph_2.0.3              R6_2.5.1                 </span></span>
<span><span class="co">#&gt; [175] labeling_0.4.3            cluster_2.1.6            </span></span>
<span><span class="co">#&gt; [177] Rhdf5lib_1.27.0           memuse_4.2-3             </span></span>
<span><span class="co">#&gt; [179] DelayedArray_0.31.9       tidyselect_1.2.1         </span></span>
<span><span class="co">#&gt; [181] vipor_0.4.7               xml2_1.3.6               </span></span>
<span><span class="co">#&gt; [183] AnnotationDbi_1.67.0      future_1.33.2            </span></span>
<span><span class="co">#&gt; [185] sfheaders_0.4.4           rsvd_1.0.5               </span></span>
<span><span class="co">#&gt; [187] munsell_0.5.1             KernSmooth_2.23-24       </span></span>
<span><span class="co">#&gt; [189] data.table_1.15.4         htmlwidgets_1.6.4        </span></span>
<span><span class="co">#&gt; [191] RColorBrewer_1.1-3        rlang_1.1.4              </span></span>
<span><span class="co">#&gt; [193] spatstat.sparse_3.1-0     spatstat.explore_3.3-1   </span></span>
<span><span class="co">#&gt; [195] ggnewscale_0.5.0          fansi_1.0.6              </span></span>
<span><span class="co">#&gt; [197] beeswarm_0.4.0</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lambda Moses, Alik Huseynov, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
